// md-editor-v3/lib/es/MdModal.mjs
import { defineComponent as x, ref as c, reactive as k, computed as $, onMounted as F, onBeforeUnmount as w, watch as v, nextTick as M, createVNode as o } from "vue";
import { p as s } from "./chunks/config.mjs";
import { g } from "./chunks/vue-tsx.mjs";
import { k as y } from "./chunks/dom.mjs";
import "@vavt/util";
const b = {
  title: {
    type: String,
    default: ""
  },
  visible: {
    type: Boolean,
    default: !1
  },
  width: {
    type: String,
    default: "auto"
  },
  height: {
    type: String,
    default: "auto"
  },
  onClose: {
    type: Function
  },
  showAdjust: {
    type: Boolean,
    default: !1
  },
  isFullscreen: {
    type: Boolean,
    default: !1
  },
  onAdjust: {
    type: Function,
    default: () => {
    }
  }
}, h = /* @__PURE__ */ x({
  name: "MdModal",
  props: b,
  emits: ["onClose"],
  setup(e, u) {
    const f = c(e.visible), n = c([`${s}-modal`]), m = c(), r = c();
    let d = () => {
    };
    const t = k({
      initPos: {
        left: "0px",
        top: "0px"
      },
      historyPos: {
        left: "0px",
        top: "0px"
      }
    }), C = $(() => e.isFullscreen ? {
      width: "100%",
      height: "100%"
    } : {
      width: e.width,
      height: e.height
    });
    return F(() => {
      d = y(r.value, (i, l) => {
        t.initPos.left = i + "px", t.initPos.top = l + "px";
      });
    }), w(() => {
      d();
    }), v(() => e.isFullscreen, (i) => {
      i ? d() : d = y(r.value, (l, a) => {
        t.initPos.left = l + "px", t.initPos.top = a + "px";
      });
    }), v(() => e.visible, (i) => {
      i ? (n.value.push("zoom-in"), f.value = i, M(() => {
        const l = m.value.offsetWidth / 2, a = m.value.offsetHeight / 2, P = document.documentElement.clientWidth / 2, p = document.documentElement.clientHeight / 2;
        t.initPos.left = P - l + "px", t.initPos.top = p - a + "px";
      }), setTimeout(() => {
        n.value = n.value.filter((l) => l !== "zoom-in");
      }, 140)) : (n.value.push("zoom-out"), setTimeout(() => {
        n.value = n.value.filter((l) => l !== "zoom-out"), f.value = i;
      }, 130));
    }), () => {
      const i = g({
        ctx: u
      }), l = g({
        props: e,
        ctx: u
      }, "title");
      return o("div", {
        style: {
          display: f.value ? "block" : "none"
        }
      }, [o("div", {
        class: `${s}-modal-mask`,
        onClick: () => {
          e.onClose ? e.onClose() : u.emit("onClose");
        }
      }, null), o("div", {
        class: n.value,
        style: {
          ...t.initPos,
          ...C.value
        },
        ref: m
      }, [o("div", {
        class: `${s}-modal-header`,
        ref: r
      }, [l || ""]), o("div", {
        class: `${s}-modal-body`
      }, [i]), o("div", {
        class: `${s}-modal-func`
      }, [e.showAdjust && o("div", {
        class: `${s}-modal-adjust`,
        onClick: (a) => {
          a.stopPropagation(), e.isFullscreen ? t.initPos = t.historyPos : (t.historyPos = t.initPos, t.initPos = {
            left: "0",
            top: "0"
          }), e.onAdjust(!e.isFullscreen);
        }
      }, [o("svg", {
        class: `${s}-icon`,
        "aria-hidden": "true"
      }, [o("use", {
        "xlink:href": `#md-editor-icon-${e.isFullscreen ? "suoxiao" : "fangda"}`
      }, null)])]), o("div", {
        class: `${s}-modal-close`,
        onClick: (a) => {
          a.stopPropagation(), e.onClose ? e.onClose() : u.emit("onClose");
        }
      }, [o("svg", {
        class: `${s}-icon`,
        "aria-hidden": "true"
      }, [o("use", {
        "xlink:href": "#md-editor-icon-close"
      }, null)])])])])]);
    };
  }
});
h.install = (e) => (e.component(h.name, h), e);
export {
  h as default
};
