// md-editor-v3/lib/es/MdEditor.mjs
var wt = Object.defineProperty;
var xt = (e, a, l) => a in e ? wt(e, a, { enumerable: !0, configurable: !0, writable: !0, value: l }) : e[a] = l;
var P = (e, a, l) => (xt(e, typeof a != "symbol" ? a + "" : a, l), l);
import { defineComponent as B, createVNode as t, inject as A, ref as U, onMounted as Z, reactive as Y, computed as fe, watch as H, nextTick as ce, Fragment as kt, cloneVNode as at, toRef as le, shallowRef as St, onBeforeUnmount as At } from "vue";
import { p as o, a as z, s as Mt, b as It, d as Lt } from "./chunks/config.mjs";
import { draggingScroll as Et, linkTo as Dt, debounce as Nt, throttle as Ht, smoothScroll as rt } from "@vavt/util";
import { a as te, D as Vt } from "./chunks/index2.mjs";
import { b as f, a as Ft } from "./chunks/index3.mjs";
import { C as Bt, a as Ot, O as Ct, T as Rt, c as Pt, b as qt, e as Ut, d as Gt, u as Kt, f as _t, g as jt, h as Wt, i as zt, j as Zt, k as Xt, M as Yt } from "./chunks/index.mjs";
import { a as Jt } from "./chunks/dom.mjs";
import pt from "./MdModal.mjs";
import { minimalSetup as Qt, EditorView as ie } from "codemirror";
import { EditorView as be, placeholder as el, keymap as tl } from "@codemirror/view";
import { languages as ll } from "@codemirror/language-data";
import { markdown as ol } from "@codemirror/lang-markdown";
import { deleteLine as al, indentWithTab as rl, undo as nl, redo as il } from "@codemirror/commands";
import { HighlightStyle as $t, syntaxHighlighting as yt, indentUnit as sl } from "@codemirror/language";
import { tags as s } from "@lezer/highlight";
import { EditorSelection as q, EditorState as ue, Compartment as cl, StateEffect as dl } from "@codemirror/state";
import { autocompletion as ul } from "@codemirror/autocomplete";
import Tt from "./MdCatalog.mjs";
import { g as nt } from "./chunks/vue-tsx.mjs";
import ml from "./NormalToolbar.mjs";
import gl from "./ModalToolbar.mjs";
import "medium-zoom";
import "copy-to-clipboard";
import "markdown-it";
import "markdown-it-image-figures";
import "markdown-it-task-lists";
import "lru-cache";
const hl = /* @__PURE__ */ B({
  setup() {
    return () => t("div", {
      class: `${o}-divider`
    }, null);
  }
}), fl = {
  noPrettier: {
    type: Boolean
  },
  // 工具栏选择显示
  toolbars: {
    type: Array,
    default: () => []
  },
  // 工具栏选择不显示
  toolbarsExclude: {
    type: Array,
    default: () => []
  },
  setting: {
    type: Object,
    default: () => ({})
  },
  screenfull: {
    type: Object,
    default: null
  },
  screenfullJs: {
    type: String,
    default: ""
  },
  updateSetting: {
    type: Function,
    default: () => {
    }
  },
  tableShape: {
    type: Array,
    default: () => [6, 4]
  },
  defToolbars: {
    type: Object
  },
  noUploadImg: {
    type: Boolean
  },
  /**
   * 是否在工具栏下面显示对应的文字名称
   *
   * @default false
   */
  showToolbarName: {
    type: Boolean
  }
}, bl = (e) => {
  var c, g, m, b;
  const a = A("editorId");
  let l = (g = (c = z.editorExtensions) == null ? void 0 : c.screenfull) == null ? void 0 : g.instance;
  const r = (b = (m = z.editorExtensions) == null ? void 0 : m.screenfull) == null ? void 0 : b.js, n = U(!1), i = ($) => {
    if (!l) {
      f.emit(a, "errorCatcher", {
        name: "fullscreen",
        message: "fullscreen is undefined"
      });
      return;
    }
    l.isEnabled ? (n.value = !0, ($ === void 0 ? !l.isFullscreen : $) ? l.request() : l.exit()) : console.error("browser does not support screenfull!");
  }, u = () => {
    l && l.isEnabled && l.on("change", () => {
      (n.value || e.setting.fullscreen) && (n.value = !1, e.updateSetting("fullscreen"));
    });
  }, d = () => {
    l = window.screenfull, u();
  };
  return Z(() => {
    if (u(), !l) {
      const $ = document.createElement("script");
      $.src = r || Mt, $.onload = d, $.id = `${o}-screenfull`, Jt($, "screenfull");
    }
  }), Z(() => {
    f.on(a, {
      name: Bt,
      callback: i
    });
  }), { fullscreenHandler: i };
}, vl = {
  tableShape: {
    type: Array,
    default: () => [6, 4]
  },
  onSelected: {
    type: Function,
    default: () => {
    }
  }
}, kl = /* @__PURE__ */ B({
  name: "TableShape",
  props: vl,
  setup(e) {
    const a = Y({
      x: -1,
      y: -1
    });
    return () => t("div", {
      class: `${o}-table-shape`,
      onMouseleave: () => {
        a.x = -1, a.y = -1;
      }
    }, [new Array(e.tableShape[1]).fill("").map((l, r) => t("div", {
      class: `${o}-table-shape-row`,
      key: `table-shape-row-${r}`
    }, [new Array(e.tableShape[0]).fill("").map((n, i) => t("div", {
      class: `${o}-table-shape-col`,
      key: `table-shape-col-${i}`,
      onMouseenter: () => {
        a.x = r, a.y = i;
      },
      onClick: () => {
        e.onSelected(a);
      }
    }, [t("div", {
      class: [`${o}-table-shape-col-default`, r <= a.x && i <= a.y && `${o}-table-shape-col-include`]
    }, null)]))]))]);
  }
}), Cl = kl, pl = {
  type: {
    type: String,
    default: "link"
  },
  visible: {
    type: Boolean,
    default: !1
  },
  onCancel: {
    type: Function,
    default: () => {
    }
  },
  onOk: {
    type: Function,
    default: () => {
    }
  }
}, $l = /* @__PURE__ */ B({
  props: pl,
  setup(e) {
    const a = A("usedLanguageText"), l = A("editorId"), r = fe(() => {
      var i, u;
      switch (e.type) {
        case "link":
          return (i = a.value.linkModalTips) == null ? void 0 : i.linkTitle;
        case "image":
          return (u = a.value.linkModalTips) == null ? void 0 : u.imageTitle;
        default:
          return "";
      }
    }), n = Y({
      desc: "",
      url: ""
    });
    return H(() => e.visible, (i) => {
      i || setTimeout(() => {
        n.desc = "", n.url = "";
      }, 200);
    }), () => t(pt, {
      title: r.value,
      visible: e.visible,
      onClose: e.onCancel
    }, {
      default: () => {
        var i, u, d, c, g;
        return [t("div", {
          class: `${o}-form-item`
        }, [t("label", {
          class: `${o}-label`,
          for: `link-desc-${l}`
        }, [(i = a.value.linkModalTips) == null ? void 0 : i.descLabel]), t("input", {
          placeholder: (u = a.value.linkModalTips) == null ? void 0 : u.descLabelPlaceHolder,
          class: `${o}-input`,
          id: `link-desc-${l}`,
          type: "text",
          value: n.desc,
          onChange: (m) => {
            n.desc = m.target.value;
          },
          autocomplete: "off"
        }, null)]), t("div", {
          class: `${o}-form-item`
        }, [t("label", {
          class: `${o}-label`,
          for: `link-url-${l}`
        }, [(d = a.value.linkModalTips) == null ? void 0 : d.urlLabel]), t("input", {
          placeholder: (c = a.value.linkModalTips) == null ? void 0 : c.urlLabelPlaceHolder,
          class: `${o}-input`,
          id: `link-url-${l}`,
          type: "text",
          value: n.url,
          onChange: (m) => {
            n.url = m.target.value;
          },
          autocomplete: "off"
        }, null)]), t("div", {
          class: `${o}-form-item`
        }, [t("button", {
          class: [`${o}-btn`, `${o}-btn-row`],
          type: "button",
          onClick: () => {
            e.onOk(n), n.desc = "", n.url = "";
          }
        }, [(g = a.value.linkModalTips) == null ? void 0 : g.buttonOK])])];
      }
    });
  }
}), yl = {
  visible: {
    type: Boolean,
    default: !1
  },
  onCancel: {
    type: Function,
    default: () => {
    }
  },
  onOk: {
    type: Function,
    default: () => {
    }
  }
}, Tl = /* @__PURE__ */ B({
  props: yl,
  setup(e) {
    var m, b, $;
    const a = A("usedLanguageText"), l = A("editorId");
    let r = ($ = (b = (m = z) == null ? void 0 : m.editorExtensions) == null ? void 0 : b.cropper) == null ? void 0 : $.instance;
    const n = U(), i = U(), u = U(), d = Y({
      cropperInited: !1,
      imgSelected: !1,
      imgSrc: "",
      // 是否全屏
      isFullscreen: !1
    });
    let c = null;
    H(() => e.visible, () => {
      e.visible && !d.cropperInited && (r = r || window.Cropper, n.value.onchange = () => {
        if (!r) {
          f.emit(l, "errorCatcher", {
            name: "Cropper",
            message: "Cropper is undefined"
          });
          return;
        }
        const T = n.value.files || [];
        if (d.imgSelected = !0, (T == null ? void 0 : T.length) > 0) {
          const v = new FileReader();
          v.onload = (h) => {
            d.imgSrc = h.target.result, ce(() => {
              c = new r(i.value, {
                viewMode: 2,
                preview: `.${o}-clip-preview-target`
                // aspectRatio: 16 / 9,
              });
            });
          }, v.readAsDataURL(T[0]);
        }
      });
    }), H(() => [d.imgSelected], () => {
      u.value.style = "";
    }), H(() => d.isFullscreen, () => {
      ce(() => {
        c == null || c.destroy(), u.value.style = "", i.value && (c = new r(i.value, {
          viewMode: 2,
          preview: `.${o}-clip-preview-target`
          // aspectRatio: 16 / 9,
        }));
      });
    });
    const g = () => {
      c.clear(), c.destroy(), c = null, n.value.value = "", d.imgSelected = !1;
    };
    return () => {
      var T;
      return t(pt, {
        class: `${o}-modal-clip`,
        title: (T = a.value.clipModalTips) == null ? void 0 : T.title,
        visible: e.visible,
        onClose: e.onCancel,
        showAdjust: !0,
        isFullscreen: d.isFullscreen,
        onAdjust: (v) => {
          d.isFullscreen = v;
        },
        width: "668px",
        height: "421px"
      }, {
        default: () => {
          var v, h;
          return [t("div", {
            class: `${o}-form-item ${o}-clip`
          }, [t("div", {
            class: `${o}-clip-main`
          }, [d.imgSelected ? t("div", {
            class: `${o}-clip-cropper`
          }, [t("img", {
            src: d.imgSrc,
            ref: i,
            style: {
              display: "none"
            },
            alt: ""
          }, null), t("div", {
            class: `${o}-clip-delete`,
            onClick: g
          }, [t("svg", {
            class: `${o}-icon`,
            "aria-hidden": "true"
          }, [t("use", {
            "xlink:href": "#md-editor-icon-delete"
          }, null)])])]) : t("div", {
            class: `${o}-clip-upload`,
            onClick: () => {
              n.value.click();
            }
          }, [t("svg", {
            class: `${o}-icon`,
            "aria-hidden": "true"
          }, [t("use", {
            "xlink:href": "#md-editor-icon-upload"
          }, null)])])]), t("div", {
            class: `${o}-clip-preview`
          }, [t("div", {
            class: `${o}-clip-preview-target`,
            ref: u
          }, null)])]), t("div", {
            class: `${o}-form-item`
          }, [t("button", {
            class: `${o}-btn`,
            type: "button",
            onClick: () => {
              if (c) {
                const k = c.getCroppedCanvas();
                f.emit(l, "uploadImage", [Ft(k.toDataURL("image/png"))], e.onOk), g();
              }
            }
          }, [((v = a.value.clipModalTips) == null ? void 0 : v.buttonUpload) || ((h = a.value.linkModalTips) == null ? void 0 : h.buttonOK)])]), t("input", {
            ref: n,
            accept: "image/*",
            type: "file",
            multiple: !1,
            style: {
              display: "none"
            }
          }, null)];
        }
      });
    };
  }
}), wl = {
  type: {
    type: String,
    default: "link"
  },
  linkVisible: {
    type: Boolean,
    default: !1
  },
  clipVisible: {
    type: Boolean,
    default: !1
  },
  onCancel: {
    type: Function,
    default: () => {
    }
  },
  onOk: {
    type: Function,
    default: () => {
    }
  }
}, xl = /* @__PURE__ */ B({
  props: wl,
  setup(e) {
    return () => t(kt, null, [t($l, {
      type: e.type,
      visible: e.linkVisible,
      onOk: e.onOk,
      onCancel: e.onCancel
    }, null), t(Tl, {
      visible: e.clipVisible,
      onOk: e.onOk,
      onCancel: e.onCancel
    }, null)]);
  }
}), Sl = /* @__PURE__ */ B({
  name: "MDEditorToolbar",
  props: fl,
  setup(e) {
    const a = A("editorId"), l = A("usedLanguageText"), {
      fullscreenHandler: r
    } = bl(e), n = `${a}-toolbar-wrapper`, i = U(), u = Y({
      title: !1,
      catalog: !1,
      // 图片上传下拉
      image: !1,
      // 表格预选
      table: !1,
      // mermaid
      mermaid: !1,
      katex: !1
    }), d = (v, h) => {
      f.emit(a, "replace", v, h);
    }, c = Y({
      type: "link",
      linkVisible: !1,
      clipVisible: !1
    }), g = U();
    Z(() => {
      f.on(a, {
        name: "openModals",
        callback(v) {
          c.type = v, c.linkVisible = !0;
        }
      });
    });
    const m = fe(() => {
      const v = e.toolbars.filter((w) => !e.toolbarsExclude.includes(w)), h = v.indexOf("="), k = h === -1 ? v : v.slice(0, h + 1), y = h === -1 ? [] : v.slice(h, Number.MAX_SAFE_INTEGER);
      return [k, y];
    }), b = U(), $ = () => {
      f.emit(a, "uploadImage", Array.from(b.value.files || [])), b.value.value = "";
    };
    Z(() => {
      b.value.addEventListener("change", $);
    });
    const T = (v) => {
      var h, k, y, w, x, I, p, M, R, L, E, O, V, D, _, G, re, J, j, K, Q, W, N, F, ee, ne, ve, ke, Ce, pe, $e, ye, Te, we, xe, Se, Ae, Me, Ie, Le, Ee, De, Ne, He, Ve, Fe, Be, Oe, Re, Pe, qe, Ue, Ge, Ke, _e, je, We, ze, Ze, Xe, Ye, Je, Qe, et, tt, lt, ot;
      if (It.includes(v))
        switch (v) {
          case "-":
            return t(hl, null, null);
          case "bold":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (h = l.value.toolbarTips) == null ? void 0 : h.bold,
              onClick: () => {
                d("bold");
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-bold"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(k = l.value.toolbarTips) == null ? void 0 : k.bold])]);
          case "underline":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (y = l.value.toolbarTips) == null ? void 0 : y.underline,
              onClick: () => {
                d("underline");
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-underline"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(w = l.value.toolbarTips) == null ? void 0 : w.underline])]);
          case "italic":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (x = l.value.toolbarTips) == null ? void 0 : x.italic,
              onClick: () => {
                d("italic");
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-italic"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(I = l.value.toolbarTips) == null ? void 0 : I.italic])]);
          case "strikeThrough":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (p = l.value.toolbarTips) == null ? void 0 : p.strikeThrough,
              onClick: () => {
                d("strikeThrough");
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-strike-through"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(M = l.value.toolbarTips) == null ? void 0 : M.strikeThrough])]);
          case "title":
            return t(te, {
              relative: `#${n}`,
              visible: u.title,
              onChange: (C) => {
                u.title = C;
              },
              overlay: t("ul", {
                class: `${o}-menu`,
                onClick: () => {
                  u.title = !1;
                }
              }, [t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  d("h1");
                }
              }, [(R = l.value.titleItem) == null ? void 0 : R.h1]), t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  d("h2");
                }
              }, [(L = l.value.titleItem) == null ? void 0 : L.h2]), t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  d("h3");
                }
              }, [(E = l.value.titleItem) == null ? void 0 : E.h3]), t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  d("h4");
                }
              }, [(O = l.value.titleItem) == null ? void 0 : O.h4]), t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  d("h5");
                }
              }, [(V = l.value.titleItem) == null ? void 0 : V.h5]), t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  d("h6");
                }
              }, [(D = l.value.titleItem) == null ? void 0 : D.h6])])
            }, {
              default: () => {
                var C, S;
                return [t("div", {
                  class: `${o}-toolbar-item`,
                  title: (C = l.value.toolbarTips) == null ? void 0 : C.title
                }, [t("svg", {
                  class: `${o}-icon`,
                  "aria-hidden": "true"
                }, [t("use", {
                  "xlink:href": "#md-editor-icon-title"
                }, null)]), e.showToolbarName && t("div", {
                  class: `${o}-toolbar-item-name`
                }, [(S = l.value.toolbarTips) == null ? void 0 : S.title])])];
              }
            });
          case "sub":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (_ = l.value.toolbarTips) == null ? void 0 : _.sub,
              onClick: () => {
                d("sub");
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-sub"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(G = l.value.toolbarTips) == null ? void 0 : G.sub])]);
          case "sup":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (re = l.value.toolbarTips) == null ? void 0 : re.sup,
              onClick: () => {
                d("sup");
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-sup"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(J = l.value.toolbarTips) == null ? void 0 : J.sup])]);
          case "quote":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (j = l.value.toolbarTips) == null ? void 0 : j.quote,
              onClick: () => {
                d("quote");
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-quote"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(K = l.value.toolbarTips) == null ? void 0 : K.quote])]);
          case "unorderedList":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (Q = l.value.toolbarTips) == null ? void 0 : Q.unorderedList,
              onClick: () => {
                d("unorderedList");
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-unordered-list"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(W = l.value.toolbarTips) == null ? void 0 : W.unorderedList])]);
          case "orderedList":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (N = l.value.toolbarTips) == null ? void 0 : N.orderedList,
              onClick: () => {
                d("orderedList");
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-ordered-list"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(F = l.value.toolbarTips) == null ? void 0 : F.orderedList])]);
          case "task":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (ee = l.value.toolbarTips) == null ? void 0 : ee.task,
              onClick: () => {
                d("task");
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-task"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(ne = l.value.toolbarTips) == null ? void 0 : ne.task])]);
          case "codeRow":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (ve = l.value.toolbarTips) == null ? void 0 : ve.codeRow,
              onClick: () => {
                d("codeRow");
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-code-row"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(ke = l.value.toolbarTips) == null ? void 0 : ke.codeRow])]);
          case "code":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (Ce = l.value.toolbarTips) == null ? void 0 : Ce.code,
              onClick: () => {
                d("code");
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-code"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(pe = l.value.toolbarTips) == null ? void 0 : pe.code])]);
          case "link":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: ($e = l.value.toolbarTips) == null ? void 0 : $e.link,
              onClick: () => {
                c.type = "link", c.linkVisible = !0;
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-link"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(ye = l.value.toolbarTips) == null ? void 0 : ye.link])]);
          case "image":
            return e.noUploadImg ? t("div", {
              class: `${o}-toolbar-item`,
              title: (Te = l.value.toolbarTips) == null ? void 0 : Te.image,
              onClick: () => {
                c.type = "image", c.linkVisible = !0;
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-image"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(we = l.value.toolbarTips) == null ? void 0 : we.image])]) : t(te, {
              relative: `#${n}`,
              visible: u.image,
              onChange: (C) => {
                u.image = C;
              },
              overlay: t("ul", {
                class: `${o}-menu`,
                onClick: () => {
                  u.title = !1;
                }
              }, [t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  c.type = "image", c.linkVisible = !0;
                }
              }, [(xe = l.value.imgTitleItem) == null ? void 0 : xe.link]), t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  b.value.click();
                }
              }, [(Se = l.value.imgTitleItem) == null ? void 0 : Se.upload]), t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  c.clipVisible = !0;
                }
              }, [(Ae = l.value.imgTitleItem) == null ? void 0 : Ae.clip2upload])])
            }, {
              default: () => {
                var C, S;
                return [t("div", {
                  class: `${o}-toolbar-item`,
                  title: (C = l.value.toolbarTips) == null ? void 0 : C.image
                }, [t("svg", {
                  class: `${o}-icon`,
                  "aria-hidden": "true"
                }, [t("use", {
                  "xlink:href": "#md-editor-icon-image"
                }, null)]), e.showToolbarName && t("div", {
                  class: `${o}-toolbar-item-name`
                }, [(S = l.value.toolbarTips) == null ? void 0 : S.image])])];
              }
            });
          case "table":
            return t(te, {
              relative: `#${n}`,
              visible: u.table,
              onChange: (C) => {
                u.table = C;
              },
              key: "bar-table",
              overlay: t(Cl, {
                tableShape: e.tableShape,
                onSelected: (C) => {
                  d("table", {
                    selectedShape: C
                  });
                }
              }, null)
            }, {
              default: () => {
                var C, S;
                return [t("div", {
                  class: `${o}-toolbar-item`,
                  title: (C = l.value.toolbarTips) == null ? void 0 : C.table
                }, [t("svg", {
                  class: `${o}-icon`,
                  "aria-hidden": "true"
                }, [t("use", {
                  "xlink:href": "#md-editor-icon-table"
                }, null)]), e.showToolbarName && t("div", {
                  class: `${o}-toolbar-item-name`
                }, [(S = l.value.toolbarTips) == null ? void 0 : S.table])])];
              }
            });
          case "revoke":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (Me = l.value.toolbarTips) == null ? void 0 : Me.revoke,
              onClick: () => {
                f.emit(a, "ctrlZ");
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-revoke"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(Ie = l.value.toolbarTips) == null ? void 0 : Ie.revoke])]);
          case "next":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (Le = l.value.toolbarTips) == null ? void 0 : Le.next,
              onClick: () => {
                f.emit(a, "ctrlShiftZ");
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-next"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(Ee = l.value.toolbarTips) == null ? void 0 : Ee.next])]);
          case "save":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (De = l.value.toolbarTips) == null ? void 0 : De.save,
              onClick: () => {
                f.emit(a, Ct);
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-baocun"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(Ne = l.value.toolbarTips) == null ? void 0 : Ne.save])]);
          case "prettier":
            return e.noPrettier ? "" : t("div", {
              class: `${o}-toolbar-item`,
              title: (He = l.value.toolbarTips) == null ? void 0 : He.prettier,
              onClick: () => {
                d("prettier");
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-prettier"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(Ve = l.value.toolbarTips) == null ? void 0 : Ve.prettier])]);
          case "pageFullscreen":
            return !e.setting.fullscreen && t("div", {
              class: `${o}-toolbar-item`,
              title: (Fe = l.value.toolbarTips) == null ? void 0 : Fe.pageFullscreen,
              onClick: () => {
                e.updateSetting("pageFullscreen");
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": `#md-editor-icon-${e.setting.pageFullscreen ? "suoxiao" : "fangda"}`
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(Be = l.value.toolbarTips) == null ? void 0 : Be.pageFullscreen])]);
          case "fullscreen":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (Oe = l.value.toolbarTips) == null ? void 0 : Oe.fullscreen,
              onClick: () => {
                r();
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": `#md-editor-icon-${e.setting.fullscreen ? "fullscreen-exit" : "fullscreen"}`
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(Re = l.value.toolbarTips) == null ? void 0 : Re.fullscreen])]);
          case "preview":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (Pe = l.value.toolbarTips) == null ? void 0 : Pe.preview,
              onClick: () => {
                e.updateSetting("preview");
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-preview"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(qe = l.value.toolbarTips) == null ? void 0 : qe.preview])]);
          case "htmlPreview":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (Ue = l.value.toolbarTips) == null ? void 0 : Ue.htmlPreview,
              onClick: () => {
                e.updateSetting("htmlPreview");
              }
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-coding"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(Ge = l.value.toolbarTips) == null ? void 0 : Ge.htmlPreview])]);
          case "catalog":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (Ke = l.value.toolbarTips) == null ? void 0 : Ke.catalog,
              onClick: () => {
                f.emit(a, Ot);
              },
              key: "bar-catalog"
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-catalog"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(_e = l.value.toolbarTips) == null ? void 0 : _e.catalog])]);
          case "github":
            return t("div", {
              class: `${o}-toolbar-item`,
              title: (je = l.value.toolbarTips) == null ? void 0 : je.github,
              onClick: () => Dt("https://github.com/imzbf/md-editor-v3")
            }, [t("svg", {
              class: `${o}-icon`,
              "aria-hidden": "true"
            }, [t("use", {
              "xlink:href": "#md-editor-icon-github"
            }, null)]), e.showToolbarName && t("div", {
              class: `${o}-toolbar-item-name`
            }, [(We = l.value.toolbarTips) == null ? void 0 : We.github])]);
          case "mermaid":
            return t(te, {
              relative: `#${n}`,
              visible: u.mermaid,
              onChange: (C) => {
                u.mermaid = C;
              },
              overlay: t("ul", {
                class: `${o}-menu`,
                onClick: () => {
                  u.mermaid = !1;
                }
              }, [t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  d("flow");
                }
              }, [(ze = l.value.mermaid) == null ? void 0 : ze.flow]), t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  d("sequence");
                }
              }, [(Ze = l.value.mermaid) == null ? void 0 : Ze.sequence]), t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  d("gantt");
                }
              }, [(Xe = l.value.mermaid) == null ? void 0 : Xe.gantt]), t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  d("class");
                }
              }, [(Ye = l.value.mermaid) == null ? void 0 : Ye.class]), t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  d("state");
                }
              }, [(Je = l.value.mermaid) == null ? void 0 : Je.state]), t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  d("pie");
                }
              }, [(Qe = l.value.mermaid) == null ? void 0 : Qe.pie]), t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  d("relationship");
                }
              }, [(et = l.value.mermaid) == null ? void 0 : et.relationship]), t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  d("journey");
                }
              }, [(tt = l.value.mermaid) == null ? void 0 : tt.journey])]),
              key: "bar-mermaid"
            }, {
              default: () => {
                var C, S;
                return [t("div", {
                  class: `${o}-toolbar-item`,
                  title: (C = l.value.toolbarTips) == null ? void 0 : C.mermaid
                }, [t("svg", {
                  class: `${o}-icon`,
                  "aria-hidden": "true"
                }, [t("use", {
                  "xlink:href": "#md-editor-icon-mermaid"
                }, null)]), e.showToolbarName && t("div", {
                  class: `${o}-toolbar-item-name`
                }, [(S = l.value.toolbarTips) == null ? void 0 : S.mermaid])])];
              }
            });
          case "katex":
            return t(te, {
              relative: `#${n}`,
              visible: u.katex,
              onChange: (C) => {
                u.katex = C;
              },
              overlay: t("ul", {
                class: `${o}-menu`,
                onClick: () => {
                  u.katex = !1;
                }
              }, [t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  d("katexInline");
                }
              }, [(lt = l.value.katex) == null ? void 0 : lt.inline]), t("li", {
                class: `${o}-menu-item`,
                onClick: () => {
                  d("katexBlock");
                }
              }, [(ot = l.value.katex) == null ? void 0 : ot.block])]),
              key: "bar-katex"
            }, {
              default: () => {
                var C, S;
                return [t("div", {
                  class: `${o}-toolbar-item`,
                  title: (C = l.value.toolbarTips) == null ? void 0 : C.katex
                }, [t("svg", {
                  class: `${o}-icon`,
                  "aria-hidden": "true"
                }, [t("use", {
                  "xlink:href": "#md-editor-icon-formula"
                }, null)]), e.showToolbarName && t("div", {
                  class: `${o}-toolbar-item-name`
                }, [(S = l.value.toolbarTips) == null ? void 0 : S.katex])])];
              }
            });
        }
      else if (e.defToolbars instanceof Array) {
        const C = e.defToolbars[v];
        return C ? at(C, {
          insert(de) {
            f.emit(a, "replace", "universal", {
              generate: de
            });
          }
        }) : "";
      } else if (e.defToolbars && e.defToolbars.children instanceof Array) {
        const C = e.defToolbars.children[v];
        return C ? at(C, {
          insert(de) {
            f.emit(a, "replace", "universal", {
              generate: de
            });
          }
        }) : "";
      } else
        return "";
    };
    return H(() => e.toolbars, () => {
      ce(() => {
        i.value && Et(i.value);
      });
    }, {
      immediate: !0
    }), () => {
      const v = m.value[0].map((k) => T(k)), h = m.value[1].map((k) => T(k));
      return t(kt, null, [e.toolbars.length > 0 && t("div", {
        class: `${o}-toolbar-wrapper`,
        ref: i,
        id: n
      }, [t("div", {
        class: [`${o}-toolbar`, e.showToolbarName && `${o}-stn`]
      }, [t("div", {
        class: `${o}-toolbar-left`,
        ref: g
      }, [v]), t("div", {
        class: `${o}-toolbar-right`
      }, [h])])]), t("input", {
        ref: b,
        accept: "image/*",
        type: "file",
        multiple: !0,
        style: {
          display: "none"
        }
      }, null), t(xl, {
        linkVisible: c.linkVisible,
        clipVisible: c.clipVisible,
        type: c.type,
        onCancel: () => {
          c.linkVisible = !1, c.clipVisible = !1;
        },
        onOk: (k) => {
          k && d(c.type, {
            desc: k.desc,
            url: k.url
          }), c.linkVisible = !1, c.clipVisible = !1;
        }
      }, null)]);
    };
  }
}), Al = (e, a) => {
  const l = Nt(() => {
    e.removeEventListener("scroll", r), e.addEventListener("scroll", r), a.removeEventListener("scroll", r), a.addEventListener("scroll", r);
  }, 50), r = (n) => {
    const i = e.clientHeight, u = a.clientHeight, d = e.scrollHeight, c = a.scrollHeight, g = (d - i) / (c - u);
    n.target === e ? (a.removeEventListener("scroll", r), a.scrollTo({
      top: e.scrollTop / g
      // behavior: 'smooth'
    }), l()) : (e.removeEventListener("scroll", r), e.scrollTo({
      top: a.scrollTop * g
      // behavior: 'smooth'
    }), l());
  };
  return [
    () => {
      l().finally(() => {
        e.dispatchEvent(new Event("scroll"));
      });
    },
    () => {
      e.removeEventListener("scroll", r), a.removeEventListener("scroll", r);
    }
  ];
}, Ml = (e, a, l) => {
  const { view: r } = l, n = (b) => r.lineBlockAt(r.state.doc.line(b).from).top, i = (b) => r.lineBlockAt(r.state.doc.line(b).from).height;
  let u = [];
  const d = () => {
    u = [];
    const b = Array.from(a.querySelectorAll("[data-line]")).map(
      (h) => Number(h.dataset.line) + 1
    ), { lines: $ } = r.state.doc;
    let T = 1, v = b.shift() ?? $;
    for (let h = 1; h <= $; h++)
      h === v && (T = h, v = b.shift() || $ + 1), u.push({
        start: T,
        end: v - 1
      });
  };
  let c = 0, g = 0;
  const m = Ht((b) => {
    var v, h, k, y, w, x, I;
    if (!((v = a.firstElementChild) != null && v.firstElementChild))
      return;
    const $ = r.lineBlockAtHeight(r.scrollDOM.scrollTop), { number: T } = r.state.doc.lineAt($.from);
    if (!(T > u.length) && !(r.state.doc.lines < u[u.length - 1].end))
      if (b.target === e) {
        if (g !== 0)
          return;
        c++;
        const p = u[T - 1], M = n(p.end) + i(p.end);
        let R = 0;
        const L = n(p.start), E = a.querySelector(`[data-line="${p.start - 1}"]`) || ((h = a.firstElementChild) == null ? void 0 : h.firstElementChild), O = a.querySelector(`[data-line="${p.end}"]`) || ((k = a.lastElementChild) == null ? void 0 : k.lastElementChild);
        let V = 0, D = 0;
        L === 0 ? (R = (r.scrollDOM.scrollTop - L) / (M - L), V = O.offsetTop) : M > r.scrollDOM.scrollHeight - r.scrollDOM.clientHeight ? (R = (r.scrollDOM.scrollTop - L) / (r.scrollDOM.scrollHeight - r.scrollDOM.clientHeight - L), D = E.offsetTop, V = a.scrollHeight - a.clientHeight - D + 10) : (R = (r.scrollDOM.scrollTop - L) / (M - L), D = E.offsetTop, V = O.offsetTop - D);
        const _ = D - 10 + V * R;
        rt(a, _, () => {
          c--;
        });
      } else {
        if (c !== 0)
          return;
        g++;
        const p = Array.from(
          a.querySelectorAll("[data-line]")
        ), M = a.scrollTop, R = a.scrollHeight;
        let L = p.length === 0 ? 1 : Math.ceil(
          Number(p[p.length - 1].dataset.line) * (M / R)
        ), E = (y = a.firstElementChild) == null ? void 0 : y.firstElementChild;
        for (let N = L; N >= 0; N--) {
          const F = a.querySelector(`[data-line="${N}"]`);
          if (F && F.offsetTop <= M) {
            E = F, L = N;
            break;
          }
        }
        let O = (w = a.firstElementChild) == null ? void 0 : w.firstElementChild, V = (x = a.firstElementChild) == null ? void 0 : x.lastElementChild;
        for (; p.length > 0; ) {
          const N = p.indexOf(E);
          if (N + 1 >= p.length)
            break;
          const F = p[N + 1];
          if (N === -1) {
            V = F;
            break;
          }
          const ee = E.offsetTop;
          if (ee > M) {
            if (N === 0) {
              O = E, V = F;
              break;
            }
            E = p[N - 1];
            continue;
          }
          const ne = F.offsetTop;
          if (ee <= M && ne > M) {
            O = E, V = F;
            break;
          }
          E = F;
        }
        const D = O.offsetTop, _ = V.offsetTop;
        let G = 0;
        const { start: re, end: J } = u[Number(O.dataset.line || 0)], j = n(re);
        let K = n(J);
        const Q = i(J);
        let W = 0;
        O === ((I = a.firstElementChild) == null ? void 0 : I.firstElementChild) ? (G = Math.max(M / _, 0), W = K + Q - j) : K > r.scrollDOM.scrollHeight - r.scrollDOM.clientHeight ? (G = Math.max(
          (M - D) / (a.scrollHeight - D - a.clientHeight),
          0
        ), K = n(r.state.doc.lines) + i(r.state.doc.lines), W = 8 + K - j - e.clientHeight) : (G = Math.max(
          (M - D) / (_ - D),
          0
        ), W = K + Q - j), rt(e, j + W * G, () => {
          g--;
        });
      }
  }, 10);
  return [
    () => {
      d(), e.addEventListener("scroll", m), a.addEventListener("scroll", m), e.dispatchEvent(new Event("scroll"));
    },
    () => {
      u = [], e.removeEventListener("scroll", m), a.removeEventListener("scroll", m);
    }
  ];
}, Il = (e, a, l) => {
  const r = A("editorId");
  let n = () => {
  }, i = () => {
  };
  const u = () => {
    n();
    const d = document.querySelector(".cm-scroller"), c = document.querySelector(
      `[id="${r}-preview-wrapper"][data-show="true"]`
    ), g = document.querySelector(
      `[id="${r}-html-wrapper"][data-show="true"]`
    );
    (c || g) && ([i, n] = (c ? Ml : Al)(
      d,
      c || g,
      l.value
    )), e.scrollAuto && i();
  };
  H(
    [
      a,
      le(e.setting, "preview"),
      le(e.setting, "htmlPreview"),
      le(e.setting, "fullscreen"),
      le(e.setting, "pageFullscreen")
    ],
    () => {
      ce(u);
    }
  ), H(
    () => e.scrollAuto,
    (d) => {
      d ? i() : n();
    }
  ), Z(u);
}, Ll = Il, El = (e, a, l) => {
  var m, b, $, T, v, h;
  let r = "", n = 0, i = 0, u = !0, d = !1;
  const c = a.getSelectedText(), g = (m = z.editorConfig) == null ? void 0 : m.mermaidTemplate;
  if (/^h[1-6]{1}$/.test(e)) {
    const k = e.replace(/^h(\d)/, (y, w) => new Array(Number(w)).fill("#", 0, w).join(""));
    r = `${k} ${c}`, n = k.length + 1;
  } else if (e === "prettier") {
    const k = window.prettier || (($ = (b = z.editorExtensions) == null ? void 0 : b.prettier) == null ? void 0 : $.prettierInstance), y = [
      ((T = window.prettierPlugins) == null ? void 0 : T.markdown) || ((h = (v = z.editorExtensions) == null ? void 0 : v.prettier) == null ? void 0 : h.parserMarkdownInstance)
    ];
    !k || y[0] === void 0 ? (f.emit(l.editorId, "errorCatcher", {
      name: "prettier",
      message: "prettier is undefined"
    }), r = a.getValue()) : r = k.format(a.getValue(), {
      parser: "markdown",
      plugins: y
    }), u = !1, d = !0;
  } else
    switch (e) {
      case "bold": {
        r = `**${c}**`, n = 2, i = -2;
        break;
      }
      case "underline": {
        r = `<u>${c}</u>`, n = 3, i = -4;
        break;
      }
      case "italic": {
        r = `*${c}*`, n = 1, i = -1;
        break;
      }
      case "strikeThrough": {
        r = `~~${c}~~`, n = 2, i = -2;
        break;
      }
      case "sub": {
        r = `<sub>${c}</sub>`, n = 5, i = -6;
        break;
      }
      case "sup": {
        r = `<sup>${c}</sup>`, n = 5, i = -6;
        break;
      }
      case "codeRow": {
        r = "`" + c + "`", n = 1, i = -1;
        break;
      }
      case "quote": {
        r = `> ${c}`, n = 2;
        break;
      }
      case "orderedList": {
        r = `1. ${c}`, n = 3;
        break;
      }
      case "unorderedList": {
        r = `- ${c}`, n = 2;
        break;
      }
      case "task": {
        r = `- [ ] ${c}`, n = 6;
        break;
      }
      case "code": {
        const k = l.text || c || "", y = l.mode || "language";
        r = `\`\`\`${y}
${k}
\`\`\`
`, n = 3, i = 3 + y.length - r.length;
        break;
      }
      case "table": {
        r = "|";
        const { selectedShape: k = { x: 1, y: 1 } } = l, { x: y, y: w } = k;
        for (let x = 0; x <= w; x++)
          r += " col |";
        r += `
|`;
        for (let x = 0; x <= w; x++)
          r += " - |";
        for (let x = 0; x <= y; x++) {
          r += `
|`;
          for (let I = 0; I <= w; I++)
            r += " content |";
        }
        n = 2, i = 5 - r.length;
        break;
      }
      case "link": {
        const { desc: k, url: y } = l;
        r = `[${k}](${y})`, u = !1;
        break;
      }
      case "image": {
        const { desc: k, url: y, urls: w } = l;
        w instanceof Array ? r = w.reduce((x, I) => x + `![${k}](${I})
`, "") : r = `![${k}](${y})
`, u = !1;
        break;
      }
      case "flow": {
        r = `\`\`\`mermaid
${(g == null ? void 0 : g.flow) || `flowchart TD 
  Start --> Stop`}
\`\`\`
`, n = 3, i = 10 - r.length;
        break;
      }
      case "sequence": {
        r = `\`\`\`mermaid
${(g == null ? void 0 : g.sequence) || `sequenceDiagram
  A->>B: hello!
  B-->>A: hi!
  A-)B: bye!`}
\`\`\`
`, n = 3, i = 10 - r.length;
        break;
      }
      case "gantt": {
        r = `\`\`\`mermaid
${(g == null ? void 0 : g.gantt) || `gantt
title A Gantt Diagram
dateFormat  YYYY-MM-DD
section Section
A task  :a1, 2014-01-01, 30d
Another task  :after a1, 20d`}
\`\`\`
`, n = 3, i = 10 - r.length;
        break;
      }
      case "class": {
        r = `\`\`\`mermaid
${(g == null ? void 0 : g.class) || `classDiagram
  class Animal
  Vehicle <|-- Car`}
\`\`\`
`, n = 3, i = 10 - r.length;
        break;
      }
      case "state": {
        r = `\`\`\`mermaid
${(g == null ? void 0 : g.state) || `stateDiagram-v2
  s1 --> s2`}
\`\`\`
`, n = 3, i = 10 - r.length;
        break;
      }
      case "pie": {
        r = `\`\`\`mermaid
${(g == null ? void 0 : g.pie) || `pie title Pets adopted by volunteers
  "Dogs" : 386
  "Cats" : 85
  "Rats" : 15`}
\`\`\`
`, n = 3, i = 10 - r.length;
        break;
      }
      case "relationship": {
        r = `\`\`\`mermaid
${(g == null ? void 0 : g.relationship) || `erDiagram
  CAR ||--o{ NAMED-DRIVER : allows
  PERSON ||--o{ NAMED-DRIVER : is`}
\`\`\`
`, n = 3, i = 10 - r.length;
        break;
      }
      case "journey": {
        r = `\`\`\`mermaid
${(g == null ? void 0 : g.journey) || `journey
  title My working day
  section Go to work
    Make tea: 5: Me
    Go upstairs: 3: Me
    Do work: 1: Me, Cat
  section Go home
    Go downstairs: 5: Me
    Sit down: 5: Me`}
\`\`\`
`, n = 3, i = 10 - r.length;
        break;
      }
      case "katexInline": {
        r = "$$", n = 1, i = -1;
        break;
      }
      case "katexBlock": {
        r = `$$

$$
`, n = 3, i = -4;
        break;
      }
      case "universal": {
        const { generate: k } = l, y = k(c);
        r = y.targetValue, u = y.select, n = y.deviationStart, i = y.deviationEnd;
      }
    }
  return {
    text: r,
    options: {
      // 是否选中
      select: u,
      // 选中时，开始位置的偏移量
      deviationStart: n,
      // 结束的偏移量
      deviationEnd: i,
      // 是否整个替换
      replaceAll: d
    }
  };
}, Dl = "#e5c07b", it = "var(--md-color)", Nl = "#56b6c2", Hl = "#ffffff", oe = "var(--md-color)", st = "#e5c07b", Vl = "#e5c07b", Fl = "var(--md-color)", ct = "#d19a66", Bl = "#c678dd", Ol = "#21252b", Rl = "#2c313a", dt = "var(--md-bk-color)", me = "var(--md-bk-color)", Pl = "#ceedfa33", ut = "#528bff", ql = be.theme(
  {
    "&": {
      color: oe,
      backgroundColor: dt
    },
    ".cm-content": {
      caretColor: ut
    },
    ".cm-cursor, .cm-dropCursor": { borderLeftColor: ut },
    "&.cm-focused > .cm-scroller > .cm-selectionLayer .cm-selectionBackground, .cm-selectionBackground, .cm-content ::selection": { backgroundColor: Pl },
    ".cm-panels": { backgroundColor: Ol, color: oe },
    ".cm-panels.cm-panels-top": { borderBottom: "2px solid black" },
    ".cm-panels.cm-panels-bottom": { borderTop: "2px solid black" },
    ".cm-searchMatch": {
      backgroundColor: "#72a1ff59",
      outline: "1px solid #457dff"
    },
    ".cm-searchMatch.cm-searchMatch-selected": {
      backgroundColor: "#6199ff2f"
    },
    ".cm-activeLine": { backgroundColor: "#ceedfa33" },
    ".cm-selectionMatch": { backgroundColor: "#aafe661a" },
    "&.cm-focused .cm-matchingBracket, &.cm-focused .cm-nonmatchingBracket": {
      backgroundColor: "#bad0f847"
    },
    ".cm-gutters": {
      backgroundColor: dt,
      color: oe,
      borderRight: "1px solid",
      borderColor: "var(--md-border-color)"
    },
    ".cm-activeLineGutter": {
      backgroundColor: Rl
    },
    ".cm-foldPlaceholder": {
      backgroundColor: "transparent",
      border: "none",
      color: "#ddd"
    },
    ".cm-tooltip": {
      border: "1px solid var(--md-border-color)",
      backgroundColor: me
    },
    ".cm-tooltip .cm-tooltip-arrow:before": {
      borderTopColor: "transparent",
      borderBottomColor: "transparent"
    },
    ".cm-tooltip .cm-tooltip-arrow:after": {
      borderTopColor: me,
      borderBottomColor: me
    },
    ".cm-tooltip-autocomplete": {
      "& > ul > li[aria-selected]": {
        color: oe
      }
    }
  },
  { dark: !0 }
), Ul = $t.define([
  { tag: s.keyword, color: Bl },
  { tag: [s.name, s.deleted, s.character, s.propertyName, s.macroName], color: it },
  { tag: [s.function(s.variableName), s.labelName], color: Vl },
  { tag: [s.color, s.constant(s.name), s.standard(s.name)], color: ct },
  { tag: [s.definition(s.name), s.separator], color: oe },
  {
    tag: [
      s.typeName,
      s.className,
      s.number,
      s.changed,
      s.annotation,
      s.modifier,
      s.self,
      s.namespace
    ],
    color: Dl
  },
  {
    tag: [
      s.operator,
      s.operatorKeyword,
      s.url,
      s.escape,
      s.regexp,
      s.link,
      s.special(s.string)
    ],
    color: Nl
  },
  { tag: [s.meta, s.comment], color: st },
  { tag: s.strong, fontWeight: "bold" },
  { tag: s.emphasis, fontStyle: "italic" },
  { tag: s.strikethrough, textDecoration: "line-through" },
  { tag: s.link, color: st, textDecoration: "underline" },
  { tag: s.heading, fontWeight: "bold", color: it },
  { tag: [s.atom, s.bool, s.special(s.variableName)], color: ct },
  { tag: [s.processingInstruction, s.string, s.inserted], color: Fl },
  { tag: s.invalid, color: Hl }
]), Gl = [
  ql,
  yt(Ul)
], Kl = "#e5c07b", mt = "#3f4a54", _l = "#56b6c2", jl = "#fff", ae = "#3f4a54", gt = "#2d8cf0", Wl = "#2d8cf0", zl = "#3f4a54", ht = "#d19a66", Zl = "#c678dd", Xl = "#21252b", Yl = "#ceedfa33", ft = "var(--md-bk-color)", ge = "var(--md-bk-color)", Jl = "#bad5fa", bt = "#3f4a54", Ql = be.theme(
  {
    "&": {
      color: ae,
      backgroundColor: ft
    },
    ".cm-content": {
      caretColor: bt
    },
    ".cm-cursor, .cm-dropCursor": { borderLeftColor: bt },
    "&.cm-focused > .cm-scroller > .cm-selectionLayer .cm-selectionBackground, .cm-selectionBackground, .cm-content ::selection": { backgroundColor: Jl },
    ".cm-panels": { backgroundColor: Xl, color: ae },
    ".cm-panels.cm-panels-top": { borderBottom: "2px solid black" },
    ".cm-panels.cm-panels-bottom": { borderTop: "2px solid black" },
    ".cm-searchMatch": {
      backgroundColor: "#72a1ff59",
      outline: "1px solid #457dff"
    },
    ".cm-searchMatch.cm-searchMatch-selected": {
      backgroundColor: "#6199ff2f"
    },
    ".cm-activeLine": { backgroundColor: "#ceedfa33" },
    ".cm-selectionMatch": { backgroundColor: "#aafe661a" },
    "&.cm-focused .cm-matchingBracket, &.cm-focused .cm-nonmatchingBracket": {
      backgroundColor: "#bad0f847"
    },
    ".cm-gutters": {
      backgroundColor: ft,
      color: ae,
      borderRight: "1px solid",
      borderColor: "var(--md-border-color)"
    },
    ".cm-activeLineGutter": {
      backgroundColor: Yl
    },
    ".cm-foldPlaceholder": {
      backgroundColor: "transparent",
      border: "none",
      color: "#ddd"
    },
    ".cm-tooltip": {
      border: "1px solid var(--md-border-color)",
      backgroundColor: ge
    },
    ".cm-tooltip .cm-tooltip-arrow:before": {
      borderTopColor: "transparent",
      borderBottomColor: "transparent"
    },
    ".cm-tooltip .cm-tooltip-arrow:after": {
      borderTopColor: ge,
      borderBottomColor: ge
    },
    ".cm-tooltip-autocomplete": {
      "& > ul > li[aria-selected]": {
        color: ae
      }
    }
  }
  // { dark: true }
), eo = $t.define([
  { tag: s.keyword, color: Zl },
  { tag: [s.name, s.deleted, s.character, s.propertyName, s.macroName], color: mt },
  { tag: [s.function(s.variableName), s.labelName], color: Wl },
  { tag: [s.color, s.constant(s.name), s.standard(s.name)], color: ht },
  { tag: [s.definition(s.name), s.separator], color: ae },
  {
    tag: [
      s.typeName,
      s.className,
      s.number,
      s.changed,
      s.annotation,
      s.modifier,
      s.self,
      s.namespace
    ],
    color: Kl
  },
  {
    tag: [
      s.operator,
      s.operatorKeyword,
      s.url,
      s.escape,
      s.regexp,
      s.link,
      s.special(s.string)
    ],
    color: _l
  },
  { tag: [s.meta, s.comment], color: gt },
  { tag: s.strong, fontWeight: "bold" },
  { tag: s.emphasis, fontStyle: "italic" },
  { tag: s.strikethrough, textDecoration: "line-through" },
  { tag: s.link, color: gt, textDecoration: "underline" },
  { tag: s.heading, fontWeight: "bold", color: mt },
  { tag: [s.atom, s.bool, s.special(s.variableName)], color: ht },
  { tag: [s.processingInstruction, s.string, s.inserted], color: zl },
  { tag: s.invalid, color: jl }
]), to = [
  Ql,
  yt(eo)
], se = (e, a, l, r, n) => (i, u, d, c) => {
  const g = `${e}${a}`.slice(c - d);
  i.dispatch(i.state.replaceSelection(`${g}${l}${r}`));
  const m = d + u.label.length + (n === "title" ? l.length : 0);
  i.dispatch({
    selection: q.create(
      [
        q.range(
          d + u.label.length + (n === "title" ? 1 : -a.length),
          m
        ),
        q.cursor(m)
      ],
      1
    )
  }), i.focus();
}, vt = (e) => (a, l, r, n) => {
  const i = e.slice(n - r);
  a.dispatch(a.state.replaceSelection(`${i} `));
}, lo = (e) => {
  const a = (l) => {
    const r = l.matchBefore(
      /^#+|^-\s*\[*\s*\]*|`+|\[|!\[*|^\|\s?\|?|^\$\$?|!+\s*\w*/
    );
    return r === null || r.from == r.to && l.explicit ? null : {
      from: r.from,
      options: [
        // 标题
        ...["h1", "h2", "h3", "h4", "h5", "h6"].map((n, i) => {
          const u = new Array(i + 1).fill("#").join("");
          return {
            label: u,
            type: "text",
            apply: vt(u)
          };
        }),
        // 任务列表
        ...["unchecked", "checked"].map((n) => {
          const i = n === "checked" ? "- [x]" : "- [ ]";
          return {
            label: i,
            type: "text",
            apply: vt(i)
          };
        }),
        // 代码
        ...[
          ["`", ""],
          ["```", "language"],
          ["```mermaid\n", ""]
        ].map((n) => ({
          label: `${n[0]}${n[1]}`,
          type: "text",
          apply: se(n[0], n[1], "", n[0] === "`" ? "`" : "\n```", "type")
        })),
        // 链接
        {
          label: "[]()",
          type: "text"
        },
        {
          label: "![]()",
          type: "text"
        },
        // 表格
        {
          label: "| |",
          type: "text",
          detail: "table",
          apply: `| col | col | col |
| - | - | - |
| content | content | content |
| content | content | content |`
        },
        // 公式
        {
          label: "$",
          type: "text",
          apply: se("$", "", "", "$", "type")
        },
        {
          label: "$$",
          type: "text",
          apply: se("$$", "", `
`, `
$$`, "title")
        },
        // 那啥？
        ...[
          "note",
          "abstract",
          "info",
          "tip",
          "success",
          "question",
          "warning",
          "failure",
          "danger",
          "bug",
          "example",
          "quote",
          "hint",
          "caution",
          "error",
          "attention"
        ].map((n) => ({
          label: `!!! ${n}`,
          type: "text",
          apply: se("!!!", ` ${n}`, " Title", `

!!!`, "title")
        }))
      ]
    };
  };
  return ul({
    override: e ? [a, ...e] : [a]
  });
}, oo = lo, X = (e) => {
  const a = new cl();
  return (r) => (a.get(e.state) ? e.dispatch({ effects: a.reconfigure(r) }) : e.dispatch({
    effects: dl.appendConfig.of(a.of(r))
  }), !0);
};
class ao {
  constructor(a) {
    P(this, "view");
    P(this, "maxLength", Number.MAX_SAFE_INTEGER);
    // 切换tabSize的执行方法。切换时，Compartment实例需要相同
    P(this, "toggleTabSize");
    P(this, "togglePlaceholder");
    /**
     * 设置全部的扩展
     */
    P(this, "setExtensions");
    P(this, "toggleDisabled");
    P(this, "toggleReadOnly");
    P(this, "toggleMaxlength");
    this.view = a, this.toggleTabSize = X(this.view), this.togglePlaceholder = X(this.view), this.setExtensions = X(this.view), this.toggleDisabled = X(this.view), this.toggleReadOnly = X(this.view), this.toggleMaxlength = X(this.view);
  }
  getValue() {
    return this.view.state.doc.toString();
  }
  /**
   * 设置内容
   *
   * @param insert 待插入内容
   * @param from 插入开始位置
   * @param to 插入结束位置
   */
  setValue(a, l = 0, r = this.view.state.doc.length) {
    this.view.dispatch({
      changes: {
        from: l,
        to: r,
        insert: a
      }
    });
  }
  /**
   * 获取选中的文本
   */
  getSelectedText() {
    const { from: a, to: l } = this.view.state.selection.main;
    return this.view.state.sliceDoc(a, l);
  }
  /**
   * 使用新的内容替换选中的内容
   *
   * @param text 待替换内容
   * @param options 替换后是否选中
   */
  replaceSelectedText(a, l = {
    // 是否选中
    select: !0,
    // 选中时，开始位置的偏移量
    deviationStart: 0,
    // 结束的偏移量
    deviationEnd: 0,
    // 直接替换所有文本
    replaceAll: !1
  }, r) {
    try {
      if (l.replaceAll) {
        if (this.setValue(a), a.length > this.maxLength)
          throw new Error("The input text is too long");
        return;
      }
      if (this.view.state.doc.length - this.getSelectedText().length + a.length > this.maxLength)
        throw new Error("The input text is too long");
      const { from: n } = this.view.state.selection.main;
      if (this.view.dispatch(this.view.state.replaceSelection(a)), l.select) {
        const i = n + a.length + l.deviationEnd;
        this.view.dispatch({
          selection: q.create(
            [
              q.range(n + l.deviationStart, i),
              q.cursor(i)
            ],
            1
          )
        });
      }
      this.view.focus();
    } catch (n) {
      if (n.message === "The input text is too long")
        f.emit(r, "errorCatcher", {
          name: "overlength",
          message: n.message,
          data: a
        });
      else
        throw n;
    }
  }
  /**
   * 设置tabSize
   *
   * @param tabSize 需要切换的大小
   */
  setTabSize(a) {
    this.toggleTabSize([
      ue.tabSize.of(a),
      sl.of(" ".repeat(a))
    ]);
  }
  /**
   * 设置placeholder
   *
   * @param t 目标内容
   */
  setPlaceholder(a) {
    this.togglePlaceholder(el(a));
  }
  focus(a) {
    if (this.view.focus(), !a)
      return;
    let l = 0, r = 0, n = 0;
    switch (a) {
      case "start":
        break;
      case "end": {
        l = r = n = this.getValue().length;
        break;
      }
      default:
        l = a.rangeAnchor || a.cursorPos, r = a.rangeHead || a.cursorPos, n = a.cursorPos;
    }
    this.view.dispatch({
      scrollIntoView: !0,
      selection: q.create(
        [q.range(l, r), q.cursor(n)],
        1
      )
    });
  }
  setDisabled(a) {
    this.toggleDisabled([be.editable.of(!a)]);
  }
  setReadOnly(a) {
    this.toggleReadOnly([ue.readOnly.of(a)]);
  }
  setMaxLength(a) {
    this.maxLength = a, this.toggleMaxlength([
      ue.changeFilter.of((l) => l.newDoc.length <= a)
    ]);
  }
}
const ro = (e) => {
  const a = A("editorId");
  return (r) => {
    if (!r.clipboardData)
      return;
    if (r.clipboardData.files.length > 0) {
      const { files: i } = r.clipboardData;
      f.emit(
        a,
        "uploadImage",
        Array.from(i).filter((u) => /image\/.*/.test(u.type))
      ), r.preventDefault();
      return;
    }
    if (e.autoDetectCode && r.clipboardData.types.includes("vscode-editor-data")) {
      const i = JSON.parse(r.clipboardData.getData("vscode-editor-data"));
      f.emit(a, "replace", "code", {
        mode: i.mode,
        text: r.clipboardData.getData("text/plain")
      }), r.preventDefault();
      return;
    }
    const n = r.clipboardData.getData("text/plain");
    e.maxlength && n.length + e.modelValue.length > e.maxlength && f.emit(a, "errorCatcher", {
      name: "overlength",
      message: "The input text is too long",
      data: n
    });
  };
}, no = ro, io = (e) => {
  const a = A("editorId");
  Z(() => {
    f.on(a, {
      name: Rt,
      callback(l) {
        var r;
        (r = e.value) == null || r.focus(l);
      }
    });
  });
}, so = io, co = (e, a) => [
  {
    key: "Ctrl-b",
    mac: "Cmd-b",
    run: () => (f.emit(e, "replace", "bold"), !0)
  },
  {
    key: "Ctrl-d",
    mac: "Cmd-d",
    run: al,
    preventDefault: !0
  },
  {
    key: "Ctrl-s",
    mac: "Cmd-s",
    run: (I) => (f.emit(e, Ct, I.state.doc.toString()), !0),
    shift: () => (f.emit(e, "replace", "strikeThrough"), !0)
  },
  {
    key: "Ctrl-u",
    mac: "Cmd-u",
    run: () => (f.emit(e, "replace", "underline"), !0),
    shift: () => (f.emit(e, "replace", "unorderedList"), !0)
  },
  {
    key: "Ctrl-i",
    mac: "Cmd-i",
    run: () => (f.emit(e, "replace", "italic"), !0),
    shift: () => (f.emit(e, "openModals", "image"), !0)
  },
  {
    key: "Ctrl-1",
    mac: "Cmd-1",
    run: () => (f.emit(e, "replace", "h1"), !0)
  },
  {
    key: "Ctrl-2",
    mac: "Cmd-2",
    run: () => (f.emit(e, "replace", "h2"), !0)
  },
  {
    key: "Ctrl-3",
    mac: "Cmd-3",
    run: () => (f.emit(e, "replace", "h3"), !0)
  },
  {
    key: "Ctrl-4",
    mac: "Cmd-4",
    run: () => (f.emit(e, "replace", "h4"), !0)
  },
  {
    key: "Ctrl-5",
    mac: "Cmd-5",
    run: () => (f.emit(e, "replace", "h5"), !0)
  },
  {
    key: "Ctrl-6",
    mac: "Cmd-6",
    run: () => (f.emit(e, "replace", "h6"), !0)
  },
  {
    key: "Ctrl-ArrowUp",
    mac: "Cmd-ArrowUp",
    run: () => (f.emit(e, "replace", "sup"), !0)
  },
  {
    key: "Ctrl-ArrowDown",
    mac: "Cmd-ArrowDown",
    run: () => (f.emit(e, "replace", "sub"), !0)
  },
  {
    key: "Ctrl-o",
    mac: "Cmd-o",
    run: () => (f.emit(e, "replace", "orderedList"), !0)
  },
  {
    key: "Ctrl-c",
    mac: "Cmd-c",
    shift: () => (f.emit(e, "replace", "code"), !0),
    any(I, p) {
      return (p.ctrlKey || p.metaKey) && p.altKey && p.code === "KeyC" ? (f.emit(e, "replace", "codeRow"), !0) : !1;
    }
  },
  {
    key: "Ctrl-l",
    mac: "Cmd-l",
    run: () => (f.emit(e, "openModals", "link"), !0)
  },
  {
    key: "Ctrl-f",
    mac: "Cmd-f",
    shift: () => a.noPrettier ? !1 : (f.emit(e, "replace", "prettier"), !0)
  },
  {
    any: (I, p) => (p.ctrlKey || p.metaKey) && p.altKey && p.shiftKey && p.code === "KeyT" ? (f.emit(e, "replace", "table"), !0) : !1
  }
], uo = co, mo = (e) => {
  const a = A("tabWidth"), l = A("editorId"), r = A("theme"), n = U(), i = St(), u = uo(l, e), d = no(e), c = [
    tl.of([...u, rl]),
    Qt,
    ol({ codeLanguages: ll }),
    // 横向换行
    ie.lineWrapping,
    ie.updateListener.of((m) => {
      m.docChanged && e.onChange(m.state.doc.toString());
    }),
    ie.domEventHandlers({
      paste: d,
      blur: e.onBlur,
      focus: e.onFocus,
      input: (m) => {
        e.onInput && e.onInput(m);
        const { data: b } = m;
        e.maxlength && e.modelValue.length + b.length > e.maxlength && f.emit(l, "errorCatcher", {
          name: "overlength",
          message: "The input text is too long",
          data: b
        });
      }
    })
  ], g = () => {
    const m = [
      ...c,
      r.value === "light" ? to : Gl,
      oo(e.completions)
    ];
    return z.codeMirrorExtensions(r.value, m, [
      ...u
    ]);
  };
  return Z(() => {
    const m = new ie({
      doc: e.modelValue,
      parent: n.value
    }), b = new ao(m);
    i.value = b, setTimeout(() => {
      b.setTabSize(a), b.setDisabled(e.disabled), b.setReadOnly(e.readonly), b.setExtensions(g()), e.placeholder && b.setPlaceholder(e.placeholder), typeof e.maxlength == "number" && b.setMaxLength(e.maxlength), e.autofocus && m.focus();
    }, 0), f.on(l, {
      name: "ctrlZ",
      callback() {
        nl(m);
      }
    }), f.on(l, {
      name: "ctrlShiftZ",
      callback() {
        il(m);
      }
    }), f.on(l, {
      name: "replace",
      callback($, T = {}) {
        var k;
        const { text: v, options: h } = El($, i.value, T);
        (k = i.value) == null || k.replaceSelectedText(v, h, l);
      }
    });
  }), H(
    [r, le(e, "completions")],
    () => {
      var m;
      (m = i.value) == null || m.setExtensions(g());
    },
    {
      deep: !0
    }
  ), H(
    () => e.modelValue,
    () => {
      var m, b;
      ((m = i.value) == null ? void 0 : m.getValue()) !== e.modelValue && ((b = i.value) == null || b.setValue(e.modelValue));
    }
  ), H(
    () => e.placeholder,
    () => {
      var m;
      (m = i.value) == null || m.setPlaceholder(e.placeholder);
    }
  ), H(
    () => e.disabled,
    () => {
      var m;
      (m = i.value) == null || m.setDisabled(e.disabled);
    }
  ), H(
    () => e.readonly,
    () => {
      var m;
      (m = i.value) == null || m.setDisabled(e.readonly);
    }
  ), H(
    () => e.maxlength,
    () => {
      var m;
      e.maxlength && ((m = i.value) == null || m.setMaxLength(e.maxlength));
    }
  ), so(i), {
    inputWrapperRef: n,
    codeMirrorUt: i
  };
}, go = mo, ho = /* @__PURE__ */ B({
  name: "MDEditorContent",
  props: Pt,
  setup(e) {
    const a = A("editorId"), l = U(""), {
      inputWrapperRef: r,
      codeMirrorUt: n
    } = go(e);
    return Ll(e, l, n), () => t("div", {
      class: `${o}-content`
    }, [t("div", {
      class: `${o}-input-wrapper`,
      ref: r
    }, null), t(qt, {
      modelValue: e.modelValue,
      setting: e.setting,
      onHtmlChanged: (i) => {
        l.value = i, e.onHtmlChanged(i);
      },
      onGetCatalog: e.onGetCatalog,
      mdHeadingId: e.mdHeadingId,
      noMermaid: e.noMermaid,
      sanitize: e.sanitize,
      noKatex: e.noKatex,
      formatCopiedText: e.formatCopiedText,
      noHighlight: e.noHighlight,
      noImgZoomIn: e.noImgZoomIn
    }, null), e.catalogVisible && t(Tt, {
      theme: e.theme,
      class: `${o}-catalog-editor`,
      editorId: a,
      mdHeadingId: e.mdHeadingId,
      key: "internal-catalog"
    }, null)]);
  }
}), fo = /* @__PURE__ */ B({
  props: {
    modelValue: {
      type: String,
      default: ""
    }
  },
  setup(e) {
    const a = A("usedLanguageText");
    return () => {
      var l, r;
      return t("div", {
        class: `${o}-footer-item`
      }, [t("label", {
        class: `${o}-footer-label`
      }, [`${(l = a.value.footer) == null ? void 0 : l.markdownTotal}:`]), t("span", null, [((r = e.modelValue) == null ? void 0 : r.length) || 0])]);
    };
  }
}), bo = {
  checked: {
    type: Boolean,
    default: !1
  },
  onChange: {
    type: Function,
    default: () => {
    }
  }
}, vo = /* @__PURE__ */ B({
  props: bo,
  setup(e) {
    return () => t("div", {
      class: [`${o}-checkbox`, e.checked && `${o}-checkbox-checked`],
      onClick: () => {
        e.onChange(!e.checked);
      }
    }, null);
  }
}), ko = {
  scrollAuto: {
    type: Boolean
  },
  onScrollAutoChange: {
    type: Function,
    default: () => {
    }
  }
}, Co = /* @__PURE__ */ B({
  props: ko,
  setup(e) {
    const a = A("usedLanguageText");
    return () => {
      var l;
      return t("div", {
        class: `${o}-footer-item`
      }, [t("label", {
        class: `${o}-footer-label`,
        onClick: () => {
          e.onScrollAutoChange(!e.scrollAuto);
        }
      }, [(l = a.value.footer) == null ? void 0 : l.scrollAuto]), t(vo, {
        checked: e.scrollAuto,
        onChange: e.onScrollAutoChange
      }, null)]);
    };
  }
}), po = {
  modelValue: {
    type: String,
    default: ""
  },
  footers: {
    type: Array,
    default: []
  },
  scrollAuto: {
    type: Boolean
  },
  onScrollAutoChange: {
    type: Function,
    default: () => {
    }
  },
  defFooters: {
    type: Object
  }
}, $o = /* @__PURE__ */ B({
  name: "MDEditorFooter",
  props: po,
  setup(e) {
    const a = fe(() => {
      const r = e.footers.indexOf("="), n = r === -1 ? e.footers : e.footers.slice(0, r), i = r === -1 ? [] : e.footers.slice(r, Number.MAX_SAFE_INTEGER);
      return [n, i];
    }), l = (r) => {
      if (Lt.includes(r))
        switch (r) {
          case "markdownTotal":
            return t(fo, {
              modelValue: e.modelValue
            }, null);
          case "scrollSwitch":
            return t(Co, {
              scrollAuto: e.scrollAuto,
              onScrollAutoChange: e.onScrollAutoChange
            }, null);
        }
      else
        return e.defFooters instanceof Array ? e.defFooters[r] || "" : e.defFooters && e.defFooters.children instanceof Array && e.defFooters.children[r] || "";
    };
    return () => {
      const r = a.value[0].map((i) => l(i)), n = a.value[1].map((i) => l(i));
      return t("div", {
        class: `${o}-footer`
      }, [t("div", {
        class: `${o}-footer-left`
      }, [r]), t("div", {
        class: `${o}-footer-right`
      }, [n])]);
    };
  }
}), yo = /* @__PURE__ */ B({
  name: "MdEditorV3",
  props: Ut,
  emits: Gt,
  setup(e, a) {
    const {
      editorId: l,
      noKatex: r,
      noMermaid: n,
      noPrettier: i,
      noUploadImg: u,
      noHighlight: d
    } = e, c = Y({
      scrollAuto: e.scrollAuto
    });
    Kt(e, a), _t(e), jt(e), Wt(e, a);
    const [g, m] = zt(e, a), b = Zt(e);
    return At(() => {
      f.clear(l);
    }), Xt(e, a, b, g, m), () => {
      var v;
      const $ = nt({
        props: e,
        ctx: a
      }, "defToolbars"), T = nt({
        props: e,
        ctx: a
      }, "defFooters");
      return t("div", {
        id: l,
        class: [o, e.class, e.theme === "dark" && `${o}-dark`, g.fullscreen || g.pageFullscreen ? `${o}-fullscreen` : ""],
        style: e.style
      }, [t(Sl, {
        noPrettier: i,
        toolbars: e.toolbars,
        toolbarsExclude: e.toolbarsExclude,
        setting: g,
        updateSetting: m,
        tableShape: e.tableShape,
        defToolbars: $,
        noUploadImg: u,
        showToolbarName: e.showToolbarName
      }, null), t(ho, {
        modelValue: e.modelValue,
        setting: g,
        mdHeadingId: e.mdHeadingId,
        noMermaid: n,
        noPrettier: i,
        sanitize: e.sanitize,
        placeholder: e.placeholder,
        noKatex: r,
        scrollAuto: c.scrollAuto,
        formatCopiedText: e.formatCopiedText,
        autofocus: e.autoFocus,
        disabled: e.disabled,
        readonly: e.readOnly,
        maxlength: e.maxLength,
        autoDetectCode: e.autoDetectCode,
        noHighlight: d,
        onChange: (h) => {
          e.onChange ? e.onChange(h) : (a.emit("update:modelValue", h), a.emit("onChange", h));
        },
        onHtmlChanged: (h) => {
          e.onHtmlChanged ? e.onHtmlChanged(h) : a.emit("onHtmlChanged", h);
        },
        onGetCatalog: (h) => {
          e.onGetCatalog ? e.onGetCatalog(h) : a.emit("onGetCatalog", h);
        },
        onBlur: (h) => {
          e.onBlur ? e.onBlur(h) : a.emit("onBlur", h);
        },
        onFocus: (h) => {
          e.onFocus ? e.onFocus(h) : a.emit("onFocus", h);
        },
        onInput: (h) => {
          e.onInput ? e.onInput(h) : a.emit("onInput", h);
        },
        completions: e.completions,
        catalogVisible: b.value,
        theme: e.theme,
        noImgZoomIn: e.noImgZoomIn
      }, null), ((v = e.footers) == null ? void 0 : v.length) > 0 && t($o, {
        modelValue: e.modelValue,
        footers: e.footers,
        defFooters: T,
        scrollAuto: c.scrollAuto,
        onScrollAutoChange: (h) => c.scrollAuto = h
      }, null)]);
    };
  }
}), he = yo;
he.install = (e) => (e.component(he.name, he), e.use(ml).use(Vt).use(gl).use(Tt).use(Yt), e);
export {
  he as default
};
