// md-editor-v3/lib/es/chunks/index.mjs
import { inject as T, onMounted as I, watch as $, toRef as W, nextTick as A, shallowRef as x, ref as B, computed as M, defineComponent as J, createVNode as H, Fragment as ie, reactive as X, provide as _, onBeforeUnmount as ce } from "vue";
import { p as h, a as P, m as se, k as N, e as j, h as ue, f as O, i as de, g as V, j as U, l as me, b as fe, d as pe } from "./config.mjs";
import { u as he, g as ge, b as g } from "./index3.mjs";
import { a as F, u as ve } from "./dom.mjs";
import ye from "medium-zoom";
import { debounce as Q } from "@vavt/util";
import be from "copy-to-clipboard";
import we from "markdown-it";
import ke from "markdown-it-image-figures";
import $e from "markdown-it-task-lists";
import Ce from "lru-cache";
const Y = "onSave", ee = "changeCatalogVisible", Ie = "changeFullscreen", G = "pageFullscreenChanged", D = "fullscreenChanged", z = "previewChanged", K = "htmlPreviewChanged", q = "catalogVisibleChanged", Se = "textareaFocus", Ee = (e, a) => {
  const l = T("editorId"), { noImgZoomIn: t } = e, n = Q(() => {
    const r = document.querySelectorAll(`#${l}-preview img`);
    r.length !== 0 && ye(r, {
      background: "#00000073"
    });
  });
  I(() => {
    !t && n();
  }), $([a, W(e.setting, "preview")], () => {
    !t && n();
  });
}, Pe = Ee, Te = (e, a) => {
  const l = T("editorId"), t = T("usedLanguageText"), n = () => {
    document.querySelectorAll(`#${l}-preview pre`).forEach((c) => {
      var m, b;
      let u = -1;
      (m = c.querySelector(".copy-button")) == null || m.remove();
      const s = ((b = t.value.copyCode) == null ? void 0 : b.text) || "复制代码", o = document.createElement("span");
      o.setAttribute("class", "copy-button"), o.dataset.tips = s, o.innerHTML = `<svg class="${h}-icon" aria-hidden="true"><use xlink:href="#${h}-icon-copy"></use></svg>`, o.addEventListener("click", () => {
        var k, v;
        clearTimeout(u);
        const y = c.querySelector("code").innerText, d = be(e.formatCopiedText(y)), p = ((k = t.value.copyCode) == null ? void 0 : k.successTips) || "已复制！", f = ((v = t.value.copyCode) == null ? void 0 : v.failTips) || "已复制！";
        o.dataset.tips = d ? p : f, u = window.setTimeout(() => {
          o.dataset.tips = s;
        }, 1500);
      }), c.appendChild(o);
    });
  }, r = () => {
    A(n);
  }, i = (c) => {
    c && A(n);
  };
  $(() => a.value, r), $(() => e.setting.preview, i), $(() => e.setting.htmlPreview, i), $(() => t.value, n), I(n);
}, Fe = Te, _e = (e) => {
  var r;
  const a = (r = P.editorExtensions) == null ? void 0 : r.highlight, l = a == null ? void 0 : a.instance, t = T("highlight"), n = x(l);
  return I(() => {
    if (!e.noHighlight && !n.value) {
      const i = document.createElement("script");
      i.src = t.value.js, i.onload = () => {
        n.value = window.hljs;
      }, i.id = `${h}-hljs`, F(i, "hljs");
      const c = document.createElement("link");
      c.rel = "stylesheet", c.href = t.value.css, c.id = `${h}-hlCss`, F(c);
    }
  }), $(
    () => t.value.css,
    (i) => {
      ve(`${h}-hlCss`, "href", i);
    }
  ), n;
}, He = _e, Me = (e) => {
  const a = T("theme"), { editorExtensions: l } = P, t = l == null ? void 0 : l.mermaid, n = x(t == null ? void 0 : t.instance), r = x(-1), i = new Ce({
    max: 1e3,
    // 缓存10分钟
    ttl: 6e5
  }), c = () => {
    const s = n.value;
    !e.noMermaid && s && (s.initialize({
      startOnLoad: !1,
      theme: a.value === "dark" ? "dark" : "default"
    }), r.value = r.value + 1);
  };
  return $(
    () => a.value,
    () => {
      i.clear(), c();
    }
  ), I(() => {
    if (!e.noMermaid && !(t != null && t.instance)) {
      const s = (t == null ? void 0 : t.js) || se;
      if (/\.mjs/.test(s))
        import(
          /* @vite-ignore */
          /* webpackIgnore: true */
          s
        ).then((o) => {
          n.value = o.default, c();
        });
      else {
        const o = document.createElement("script");
        o.id = `${h}-mermaid`, o.src = s, o.onload = () => {
          n.value = window.mermaid, c();
        }, F(o, "mermaid");
      }
    }
  }), { mermaidRef: n, reRenderRef: r, replaceMermaid: () => {
    A(() => {
      if (!e.noMermaid && n.value) {
        const s = document.querySelectorAll(
          `div.${h}-mermaid`
        ), o = document.createElement("div");
        o.style.width = document.body.offsetWidth + "px", o.style.height = document.body.offsetHeight + "px", o.style.position = "fixed", o.style.zIndex = "-10000", o.style.top = "-10000";
        let m = s.length;
        m > 0 && document.body.appendChild(o), s.forEach(async (b) => {
          let y = i.get(b.innerText);
          if (!y) {
            const p = he(), f = n.value.renderAsync || n.value.render;
            let k = "";
            try {
              k = await f(p, b.innerText, o);
            } catch {
            }
            y = typeof k == "string" ? k : k.svg, i.set(b.innerText, y);
          }
          const d = document.createElement("p");
          d.className = `${h}-mermaid`, d.setAttribute("data-processed", ""), d.innerHTML = y, b.dataset.line !== void 0 && (d.dataset.line = b.dataset.line), b.replaceWith(d), --m === 0 && o.remove();
        });
      }
    });
  } };
}, xe = Me, Re = (e) => {
  var n;
  const a = (n = P.editorExtensions) == null ? void 0 : n.katex, l = a == null ? void 0 : a.instance, t = x(l);
  return I(() => {
    if (!e.noKatex && !t.value) {
      const r = document.createElement("script");
      r.src = (a == null ? void 0 : a.js) || N.js, r.onload = () => {
        t.value = window.katex;
      }, r.id = `${h}-katex`;
      const i = document.createElement("link");
      i.rel = "stylesheet", i.href = (a == null ? void 0 : a.css) || N.css, i.id = `${h}-katexCss`, F(r, "katex"), F(i);
    }
  }), t;
}, Ae = Re, Be = (e, a) => {
  const l = e.renderer.rules.fence.bind(e.renderer.rules);
  e.renderer.rules.fence = (t, n, r, i, c) => {
    const u = t[n], s = u.content.trim();
    if (u.info === "mermaid") {
      let o;
      return t[n].map && t[n].level === 0 && (o = t[n].map[0], t[n].attrSet("data-line", String(o))), `<div class="${h}-mermaid" ${o !== void 0 ? "data-line=" + o : ""} data-mermaid-theme=${a.themeRef.value}>${s}</div>`;
    }
    return l(t, n, r, i, c);
  };
}, Le = Be, Z = (e, a) => {
  let l = !0, t = !0;
  const n = e.posMax, r = a > 0 ? e.src.charCodeAt(a - 1) : -1, i = a + 1 <= n ? e.src.charCodeAt(a + 1) : -1;
  return (r === 32 || r === 9 || i >= 48 && i <= 57) && (t = !1), (i === 32 || i === 9) && (l = !1), {
    can_open: l,
    can_close: t
  };
}, Ne = (e, a) => {
  let l, t, n, r;
  if (e.src[e.pos] !== "$")
    return !1;
  if (n = Z(e, e.pos), !n.can_open)
    return a || (e.pending += "$"), e.pos += 1, !0;
  const i = e.pos + 1;
  for (l = i; (l = e.src.indexOf("$", l)) !== -1; ) {
    for (r = l - 1; e.src[r] === "\\"; )
      r -= 1;
    if ((l - r) % 2 == 1)
      break;
    l += 1;
  }
  return l === -1 ? (a || (e.pending += "$"), e.pos = i, !0) : l - i === 0 ? (a || (e.pending += "$$"), e.pos = i + 1, !0) : (n = Z(e, l), n.can_close ? (a || (t = e.push("math_inline", "math", 0), t.markup = "$", t.content = e.src.slice(i, l)), e.pos = l + 1, !0) : (a || (e.pending += "$"), e.pos = i, !0));
}, je = (e, a, l, t) => {
  let n, r, i, c, u = !1, s = e.bMarks[a] + e.tShift[a], o = e.eMarks[a];
  if (s + 2 > o || e.src.slice(s, s + 2) !== "$$")
    return !1;
  if (s += 2, n = e.src.slice(s, o), t)
    return !0;
  for (n.trim().slice(-2) === "$$" && (n = n.trim().slice(0, -2), u = !0), i = a; !u && (i++, !(i >= l || (s = e.bMarks[i] + e.tShift[i], o = e.eMarks[i], s < o && e.tShift[i] < e.blkIndent))); )
    e.src.slice(s, o).trim().slice(-2) === "$$" && (c = e.src.slice(0, o).lastIndexOf("$$"), r = e.src.slice(s, c), u = !0);
  e.line = i + 1;
  const m = e.push("math_block", "math", 0);
  return m.block = !0, m.content = (n && n.trim() ? n + `
` : "") + e.getLines(a + 1, i, e.tShift[a], !0) + (r && r.trim() ? r : ""), m.map = [a, e.line], m.markup = "$$", !0;
}, Oe = (e, a) => {
  const l = (n) => {
    if (a.katexRef.value) {
      const r = a.katexRef.value.renderToString(n, {
        throwOnError: !1
      });
      return `<span class="${h}-katex-inline" data-processed>${r}</span>`;
    } else
      return `<span class="${h}-katex-inline">${n}</span>`;
  }, t = (n, r) => {
    if (a.katexRef.value) {
      const i = a.katexRef.value.renderToString(n, {
        throwOnError: !1,
        displayMode: !0
      });
      return `<p class="${h}-katex-block" data-line=${r} data-processed>${i}</p>`;
    } else
      return `<p class="${h}-katex-block" data-line=${r}>${n}</p>`;
  };
  e.inline.ruler.after("escape", "math_inline", Ne), e.block.ruler.after("blockquote", "math_block", je, {
    alt: ["paragraph", "reference", "blockquote", "list"]
  }), e.renderer.rules.math_inline = (n, r) => l(n[r].content), e.renderer.rules.math_block = (n, r) => t(n[r].content, n[r].map[0]) + `
`;
}, Ve = Oe, Ue = (e, a) => {
  a = a || {};
  const l = 3, t = a.marker || "!", n = t.charCodeAt(0), r = t.length;
  let i = "", c = "";
  const u = (o, m, b, y, d) => {
    const p = o[m];
    return p.type === "admonition_open" ? (o[m].attrPush([
      "class",
      `${h}-admonition ${h}-admonition-${p.info}`
    ]), o[m].attrSet("data-line", String(o[m].map[0]))) : p.type === "admonition_title_open" && o[m].attrPush(["class", `${h}-admonition-title`]), d.renderToken(o, m, b);
  }, s = (o) => {
    const m = o.trim().split(" ", 2);
    c = "", i = m[0], m.length > 1 && (c = o.substring(i.length + 2)), (c === "" || !c) && (c = i);
  };
  e.block.ruler.before(
    "code",
    "admonition",
    (o, m, b, y) => {
      let d, p, f, k = !1, v = o.bMarks[m] + o.tShift[m], w = o.eMarks[m];
      if (n !== o.src.charCodeAt(v))
        return !1;
      for (d = v + 1; d <= w && t[(d - v) % r] === o.src[d]; d++)
        ;
      const C = Math.floor((d - v) / r);
      if (C !== l)
        return !1;
      d -= (d - v) % r;
      const S = o.src.slice(v, d), E = o.src.slice(d, w);
      if (s(E), y)
        return !0;
      for (p = m; p++, !(p >= b || (v = o.bMarks[p] + o.tShift[p], w = o.eMarks[p], v < w && o.sCount[p] < o.blkIndent)); )
        if (n === o.src.charCodeAt(v) && !(o.sCount[p] - o.blkIndent >= 4)) {
          for (d = v + 1; d <= w && t[(d - v) % r] === o.src[d]; d++)
            ;
          if (!(Math.floor((d - v) / r) < C) && (d -= (d - v) % r, d = o.skipSpaces(d), !(d < w))) {
            k = !0;
            break;
          }
        }
      const L = o.parentType, le = o.lineMax;
      return o.parentType = "root", o.lineMax = p, f = o.push("admonition_open", "div", 1), f.markup = S, f.block = !0, f.info = i, f.map = [m, p], f = o.push("admonition_title_open", "p", 1), f.markup = S + " " + i, f.map = [m, p], f = o.push("inline", "", 0), f.content = c, f.map = [m, o.line - 1], f.children = [], f = o.push("admonition_title_close", "p", -1), f.markup = S + " " + i, o.md.block.tokenize(o, m + 1, p), f = o.push("admonition_close", "div", -1), f.markup = o.src.slice(v, d), f.block = !0, o.parentType = L, o.lineMax = le, o.line = p + (k ? 1 : 0), !0;
    },
    {
      alt: ["paragraph", "reference", "blockquote", "list"]
    }
  ), e.renderer.rules.admonition_open = u, e.renderer.rules.admonition_title_open = u, e.renderer.rules.admonition_title_close = u, e.renderer.rules.admonition_close = u;
}, Ge = Ue, De = (e, a) => {
  e.renderer.rules.heading_open = (l, t) => {
    const n = l[t], r = l[t + 1].content, i = n.markup.length;
    return a.headsRef.value.push({
      text: r,
      level: i
    }), n.map && n.level === 0 && (n.attrSet("data-line", String(n.map[0])), n.attrSet(
      "id",
      a.mdHeadingId(r, i, a.headsRef.value.length)
    )), e.renderer.renderToken(l, t, a);
  }, e.renderer.rules.heading_close = (l, t, n, r, i) => i.renderToken(l, t, n);
}, ze = De, Ke = (e, a) => {
  const l = e.renderer.rules.fence, t = e.utils.unescapeAll, n = /\[(\w*)(?::([\w ]*))?\]/;
  function r(s) {
    return s.info ? t(s.info).trim() : "";
  }
  function i(s) {
    const o = r(s), [m = null, b = ""] = (n.exec(o) || []).slice(1);
    return [m, b];
  }
  function c(s) {
    const o = r(s);
    return o ? o.split(/(\s+)/g)[0] : "";
  }
  const u = (s, o, m, b, y) => {
    if (s[o].hidden)
      return "";
    const [d, p] = i(s[o]);
    if (d === null)
      return l(s, o, m, b, y);
    let f, k, v, w, C = "", S = "";
    for (let E = o; E < s.length && (f = s[E], [k, v] = i(f), k === d); E++)
      f.info = f.info.replace(n, ""), f.hidden = !0, w = E - o > 0 ? "" : " checked", C += `<li><input type="radio" name="label-group-${a.editorId}-${o}"${w}><label for="group-${a.editorId}-${o}-tab-${E - o}" onclick="this.previousElementSibling.click()">${v || c(f)}</label></li>
`, S += `<input type="radio" id="group-${a.editorId}-${o}-tab-${E - o}" name="group-${a.editorId}-${o}"${w}>
` + l(s, E, m, b, y);
    return `<div class="code-tabs">
<ul>
` + C + `</ul>
` + S + "</div>";
  };
  e.renderer.rules.fence = u;
}, qe = Ke, Ze = (e) => {
  [
    "paragraph_open",
    "table_open",
    "ordered_list_open",
    "bullet_list_open",
    "blockquote_open",
    "hr",
    "html_block",
    "fence"
  ].forEach((a) => {
    const l = e.renderer.rules[a];
    l ? e.renderer.rules[a] = (t, n, r, i, c) => {
      let u;
      const s = l(t, n, r, i, c);
      return t[n].map && t[n].level === 0 ? (u = t[n].map[0], s.replace(/^(<[^>]*)/, `$1 data-line="${u}"`)) : s;
    } : e.renderer.rules[a] = (t, n, r, i, c) => {
      let u;
      return t[n].map && t[n].level === 0 && (u = t[n].map[0], t[n].attrSet("data-line", String(u))), c.renderToken(t, n, r);
    };
  });
}, We = (e, a) => {
  const { editorConfig: l, markdownItConfig: t, markdownItPlugins: n } = P, r = T("editorId"), i = T("showCodeRowNumber"), c = T("theme"), u = B([]), s = He(e), o = Ae(e), { reRenderRef: m, replaceMermaid: b } = xe(e), y = we({
    html: !0,
    breaks: !0
  });
  t(y);
  const d = [
    {
      type: "katex",
      plugin: Ve,
      options: { katexRef: o }
    },
    {
      type: "image",
      plugin: ke,
      options: { figcaption: !0, classes: "md-zoom" }
    },
    {
      type: "admonition",
      plugin: Ge,
      options: {}
    },
    {
      type: "taskList",
      plugin: $e,
      options: {}
    },
    {
      type: "heading",
      plugin: ze,
      options: { mdHeadingId: e.mdHeadingId, headsRef: u }
    },
    {
      type: "codeTabs",
      plugin: qe,
      options: { editorId: r }
    }
  ];
  e.noMermaid || d.push({
    type: "mermaid",
    plugin: Le,
    options: { themeRef: c }
  }), n(d).forEach((w) => {
    y.use(w.plugin, w.options);
  }), y.set({
    highlight: (w, C) => {
      let S;
      !e.noHighlight && s.value ? s.value.getLanguage(C) ? S = s.value.highlight(w, {
        language: C,
        ignoreIllegals: !0
      }).value : S = s.value.highlightAuto(w).value : S = y.utils.escapeHtml(w);
      const E = i ? ge(S.trim()) : `<span class="code-block">${S.trim()}</span>`;
      return `<pre><code class="language-${C}" language=${C}>${E}</code></pre>`;
    }
  }), Ze(y);
  const p = B(e.sanitize(y.render(e.modelValue))), f = () => {
    g.emit(r, "buildFinished", p.value), e.onHtmlChanged(p.value), e.onGetCatalog(u.value), g.emit(r, "catalogChanged", u.value), b();
  };
  I(f);
  const k = Q(
    async () => {
      u.value = [], p.value = e.sanitize(y.render(e.modelValue)), f();
    },
    (l == null ? void 0 : l.renderDelay) !== void 0 ? l == null ? void 0 : l.renderDelay : a ? 0 : 500
  ), v = M(() => (e.noKatex || o.value) && (e.noHighlight || s.value));
  return $([W(e, "modelValue"), v, m], k), I(() => {
    g.on(r, {
      name: "pushCatalog",
      callback() {
        g.emit(r, "catalogChanged", u.value);
      }
    });
  }), { html: p };
}, Je = We, te = {
  modelValue: {
    type: String,
    default: ""
  },
  setting: {
    type: Object,
    default: () => ({})
  },
  onHtmlChanged: {
    type: Function,
    default: () => {
    }
  },
  onGetCatalog: {
    type: Function,
    default: () => {
    }
  },
  mdHeadingId: {
    type: Function,
    default: () => ""
  },
  noMermaid: {
    type: Boolean,
    default: !1
  },
  sanitize: {
    type: Function,
    default: (e) => e
  },
  // 不使用该函数功能
  noKatex: {
    type: Boolean,
    default: !1
  },
  formatCopiedText: {
    type: Function,
    default: (e) => e
  },
  noHighlight: {
    type: Boolean,
    default: !1
  },
  previewOnly: {
    type: Boolean,
    default: !1
  },
  noImgZoomIn: {
    type: Boolean
  }
}, dt = {
  ...te,
  onChange: {
    type: Function,
    default: () => {
    }
  },
  placeholder: {
    type: String,
    default: ""
  },
  scrollAuto: {
    type: Boolean
  },
  autofocus: {
    type: Boolean
  },
  disabled: {
    type: Boolean
  },
  readonly: {
    type: Boolean
  },
  maxlength: {
    type: Number
  },
  autoDetectCode: {
    type: Boolean
  },
  /**
   * 输入框失去焦点时触发事件
   */
  onBlur: {
    type: Function,
    default: () => {
    }
  },
  /**
   * 输入框获得焦点时触发事件
   */
  onFocus: {
    type: Function,
    default: () => {
    }
  },
  noPrettier: {
    type: Boolean
  },
  completions: {
    type: Array
  },
  catalogVisible: {
    type: Boolean
  },
  theme: {
    type: String,
    default: "light"
  },
  onInput: {
    type: Function
  }
}, Xe = /* @__PURE__ */ J({
  name: "ContentPreview",
  props: te,
  setup(e) {
    const a = T("editorId"), l = T("previewTheme"), t = T("showCodeRowNumber"), {
      html: n
    } = Je(e, e.previewOnly);
    return Fe(e, n), Pe(e, n), () => H(ie, null, [H("div", {
      id: `${a}-preview-wrapper`,
      class: `${h}-preview-wrapper`,
      "data-show": e.setting.preview,
      key: "content-preview-wrapper"
    }, [H("article", {
      id: `${a}-preview`,
      class: [`${h}-preview`, `${l == null ? void 0 : l.value}-theme`, t && `${h}-scrn`],
      innerHTML: n.value
    }, null)]), !e.previewOnly && H("div", {
      id: `${a}-html-wrapper`,
      class: `${h}-preview-wrapper`,
      "data-show": e.setting.htmlPreview,
      key: "html-preview-wrapper"
    }, [H("div", {
      class: `${h}-html`
    }, [n.value])])]);
  }
}), mt = (e, a) => {
  const { editorId: l } = e, t = X({
    // 是否已编译成html
    buildFinished: !1,
    // 存储当前最新的html
    html: ""
  });
  $(
    () => e.modelValue,
    () => {
      t.buildFinished = !1;
    }
  ), I(() => {
    g.on(l, {
      name: "buildFinished",
      callback(n) {
        t.buildFinished = !0, t.html = n;
      }
    }), g.on(l, {
      name: Y,
      callback() {
        const n = new Promise((r) => {
          if (t.buildFinished)
            r(t.html);
          else {
            const i = (c) => {
              r(c), g.remove(l, "buildFinished", i);
            };
            g.on(l, {
              name: "buildFinished",
              callback: i
            });
          }
        });
        e.onSave ? e.onSave(e.modelValue, n) : a.emit("onSave", e.modelValue, n);
      }
    });
  });
}, ne = (e) => {
  var n, r;
  const { editorId: a } = e, l = (r = (n = P) == null ? void 0 : n.editorExtensions) == null ? void 0 : r.highlight;
  _("editorId", a), _(
    "theme",
    M(() => e.theme)
  ), _(
    "highlight",
    M(() => {
      const i = {
        ...j,
        ...l == null ? void 0 : l.css
      }, c = e.codeStyleReverse && e.codeStyleReverseList.includes(e.previewTheme) ? "dark" : e.theme;
      return {
        js: (l == null ? void 0 : l.js) || ue,
        css: i[e.codeTheme] ? i[e.codeTheme][c] : j.atom[c]
      };
    })
  ), _("showCodeRowNumber", e.showCodeRowNumber);
  const t = M(() => {
    var c, u;
    const i = {
      ...O,
      ...(u = (c = P) == null ? void 0 : c.editorConfig) == null ? void 0 : u.languageUserDefined
    };
    return i[e.language] ? i[e.language] : O["zh-CN"];
  });
  _("usedLanguageText", t), _(
    "previewTheme",
    M(() => e.previewTheme)
  );
}, ft = (e) => {
  ne(e), _("tabWidth", e.tabWidth);
}, oe = (e) => {
  I(() => {
    var l;
    const a = document.createElement("script");
    a.src = ((l = P.editorExtensions) == null ? void 0 : l.iconfont) || de, a.id = `${h}-icon`, e.noIconfont || F(a);
  });
}, pt = (e) => {
  var c, u, s, o, m, b;
  const { noPrettier: a, noUploadImg: l } = e, { editorExtensions: t } = P, n = a || !!((u = (c = P.editorExtensions) == null ? void 0 : c.prettier) != null && u.prettierInstance), r = a || !!((o = (s = P.editorExtensions) == null ? void 0 : s.prettier) != null && o.parserMarkdownInstance), i = l || !!((b = (m = P.editorExtensions) == null ? void 0 : m.cropper) != null && b.instance);
  I(() => {
    var k, v, w, C;
    const y = document.createElement("script"), d = document.createElement("script");
    y.src = ((k = t == null ? void 0 : t.prettier) == null ? void 0 : k.standaloneJs) || V.main, y.id = `${h}-prettier`, d.src = ((v = t == null ? void 0 : t.prettier) == null ? void 0 : v.parserMarkdownJs) || V.markdown, d.id = `${h}-prettierMD`;
    const p = document.createElement("link");
    p.rel = "stylesheet", p.href = ((w = t == null ? void 0 : t.cropper) == null ? void 0 : w.css) || U.css, p.id = `${h}-cropperCss`;
    const f = document.createElement("script");
    f.src = ((C = t == null ? void 0 : t.cropper) == null ? void 0 : C.js) || U.js, f.id = `${h}-cropper`, i || (F(p), F(f)), n || F(y), r || F(d);
  }), oe(e);
}, ht = (e, a) => {
  const { editorId: l } = e;
  I(() => {
    g.on(l, {
      name: "errorCatcher",
      callback: (t) => {
        e.onError instanceof Function ? e.onError(t) : a.emit("onError", t);
      }
    });
  });
}, gt = (e, a) => {
  const { editorId: l } = e, t = X({
    pageFullscreen: e.pageFullscreen,
    fullscreen: !1,
    preview: e.preview,
    htmlPreview: e.preview ? !1 : e.htmlPreview
  }), n = (c, u) => {
    t[c] = u === void 0 ? !t[c] : u, c === "preview" && t.preview ? t.htmlPreview = !1 : c === "htmlPreview" && t.htmlPreview && (t.preview = !1);
  };
  let r = "";
  const i = () => {
    t.pageFullscreen || t.fullscreen ? document.body.style.overflow = "hidden" : document.body.style.overflow = r;
  };
  return $(() => [t.pageFullscreen, t.fullscreen], i), I(() => {
    g.on(l, {
      name: "uploadImage",
      callback(c, u) {
        const s = (o) => {
          g.emit(l, "replace", "image", {
            desc: "",
            urls: o
          }), u && u();
        };
        e.onUploadImg ? e.onUploadImg(c, s) : a.emit("onUploadImg", c, s);
      }
    }), r = document.body.style.overflow, i();
  }), [t, n];
}, vt = (e) => {
  const { editorId: a } = e, l = B(!1);
  return I(() => {
    g.on(a, {
      name: ee,
      callback: (n) => {
        n === void 0 ? l.value = !l.value : l.value = n;
      }
    });
  }), M(() => !e.toolbarsExclude.includes("catalog") && e.toolbars.includes("catalog") && l.value);
}, yt = (e, a, l, t, n) => {
  const { editorId: r } = e;
  $(
    () => t.pageFullscreen,
    (c) => {
      g.emit(r, G, c);
    }
  ), $(
    () => t.fullscreen,
    (c) => {
      g.emit(r, D, c);
    }
  ), $(
    () => t.preview,
    (c) => {
      g.emit(r, z, c);
    }
  ), $(
    () => t.htmlPreview,
    (c) => {
      g.emit(r, K, c);
    }
  ), $(l, (c) => {
    g.emit(r, q, c);
  });
  const i = {
    on(c, u) {
      switch (c) {
        case "pageFullscreen": {
          g.on(r, {
            name: G,
            callback(s) {
              u(s);
            }
          });
          break;
        }
        case "fullscreen": {
          g.on(r, {
            name: D,
            callback(s) {
              u(s);
            }
          });
          break;
        }
        case "preview": {
          g.on(r, {
            name: z,
            callback(s) {
              u(s);
            }
          });
          break;
        }
        case "htmlPreview": {
          g.on(r, {
            name: K,
            callback(s) {
              u(s);
            }
          });
          break;
        }
        case "catalog": {
          g.on(r, {
            name: q,
            callback(s) {
              u(s);
            }
          });
          break;
        }
      }
    },
    togglePageFullscreen(c) {
      n("pageFullscreen", c);
    },
    toggleFullscreen(c) {
      g.emit(r, Ie, c);
    },
    togglePreview(c) {
      n("preview", c);
    },
    toggleHtmlPreview(c) {
      n("htmlPreview", c);
    },
    toggleCatalog(c) {
      g.emit(r, ee, c);
    },
    triggerSave() {
      g.emit(r, Y);
    },
    insert(c) {
      g.emit(r, "replace", "universal", { generate: c });
    },
    focus(c) {
      g.emit(r, Se, c);
    }
  };
  a.expose(i);
}, Qe = (e) => e, re = {
  /**
   * markdown content.
   *
   * @default ''
   */
  modelValue: {
    type: String,
    default: ""
  },
  /**
   * 主题，支持light和dark
   *
   * @default 'light'
   */
  theme: {
    type: String,
    default: "light"
  },
  /**
   * 外层类名
   *
   * @default ''
   */
  class: {
    type: String,
    default: ""
  },
  /**
   * 预设语言名称
   *
   * @default 'zh-CN'
   */
  language: {
    type: String,
    default: "zh-CN"
  },
  /**
   * html变化事件
   */
  onHtmlChanged: {
    type: Function
  },
  /**
   * 获取目录结构
   */
  onGetCatalog: {
    type: Function
  },
  /**
   * 编辑器唯一标识
   *
   * @default 'md-editor-v3'
   */
  editorId: {
    type: String,
    default: me
  },
  /**
   * 预览中代码是否显示行号
   *
   * @default false
   */
  showCodeRowNumber: {
    type: Boolean,
    default: !1
  },
  /**
   * 预览内容样式
   *
   * @default 'default'
   */
  previewTheme: {
    type: String,
    default: "default"
  },
  /**
   * 编辑器样式
   */
  style: {
    type: Object,
    default: () => ({})
  },
  /**
   * 标题的id生成方式
   *
   * @default (text: string) => text
   */
  mdHeadingId: {
    type: Function,
    default: Qe
  },
  /**
   *
   * 不能保证文本正确的情况，在marked编译md文本后通过该方法处理
   * 推荐DOMPurify、sanitize-html
   *
   * @default (text: string) => text
   */
  sanitize: {
    type: Function,
    default: (e) => e
  },
  /**
   * 不使用该mermaid
   *
   * @default false
   */
  noMermaid: {
    type: Boolean,
    default: !1
  },
  /**
   * 不使用katex
   *
   * @default false
   */
  noKatex: {
    type: Boolean,
    default: !1
  },
  /**
   * 代码主题
   *
   * @default 'atom'
   */
  codeTheme: {
    type: String,
    default: "atom"
  },
  /**
   * 不插入iconfont链接
   *
   * @default false
   */
  noIconfont: {
    type: Boolean
  },
  /**
   * 复制代码格式化方法
   *
   * @default (text) => text
   */
  formatCopiedText: {
    type: Function,
    default: (e) => e
  },
  /**
   * 某些预览主题的代码模块背景是暗色系
   * 将这个属性设置为true，会自动在该主题下的light模式下使用暗色系的代码风格
   *
   * @default true
   */
  codeStyleReverse: {
    type: Boolean,
    default: !0
  },
  /**
   * 需要自动调整的预览主题
   *
   * @default ['default', 'mk-cute']
   */
  codeStyleReverseList: {
    type: Array,
    default: ["default", "mk-cute"]
  },
  noHighlight: {
    type: Boolean,
    default: !1
  },
  /**
   * 是否关闭编辑器默认的放大缩小功能
   */
  noImgZoomIn: {
    type: Boolean,
    default: !1
  }
}, bt = {
  ...re,
  /**
   * input回调事件
   */
  onChange: {
    type: Function
  },
  /**
   * input回调事件
   */
  onSave: {
    type: Function
  },
  /**
   * 上传图片事件
   */
  onUploadImg: {
    type: Function
  },
  /**
   * 是否页面内全屏
   *
   * @default false
   */
  pageFullscreen: {
    type: Boolean,
    default: !1
  },
  /**
   * 是否展开预览
   *
   * @default true
   */
  preview: {
    type: Boolean,
    default: !0
  },
  /**
   * 是否展开html预览
   *
   * @default false
   */
  htmlPreview: {
    type: Boolean,
    default: !1
  },
  /**
   * 仅预览模式，不显示toolbar和编辑框
   *
   * @4.0.0开始移除该设置，使用组件MdPreview替换
   *
   * @default false
   */
  // previewOnly: {
  //   type: Boolean as PropType<boolean>,
  //   default: false
  // },
  /**
   * 工具栏选择显示
   *
   * @default allToolbar
   */
  toolbars: {
    type: Array,
    default: fe
  },
  /**
   * 工具栏选择不显示
   *
   * @default []
   */
  toolbarsExclude: {
    type: Array,
    default: []
  },
  /**
   * 格式化md
   *
   * @default true
   */
  noPrettier: {
    type: Boolean,
    default: !1
  },
  /**
   * 一个tab等于空格数
   *
   * @default 2
   */
  tabWidth: {
    type: Number,
    default: 2
  },
  /**
   * 表格预设格子数
   *
   * @default [6, 4]
   */
  tableShape: {
    type: Array,
    default: () => [6, 4]
  },
  /**
   * 空提示
   *
   * @default ''
   */
  placeholder: {
    type: String,
    default: ""
  },
  /**
   * 自定义的工具栏列表
   */
  defToolbars: {
    type: [String, Object]
  },
  /**
   * 内部错误捕获
   */
  onError: {
    type: Function
  },
  /**
   * 页脚列表显示顺序
   */
  footers: {
    type: Array,
    default: pe
  },
  /**
   * 是否默认激活输入框和预览框同步滚动
   *
   * @default true
   */
  scrollAuto: {
    type: Boolean,
    default: !0
  },
  /**
   * 自定义的也叫工具组件列表
   */
  defFooters: {
    type: [String, Object]
  },
  /**
   * 是否禁用上传图片
   *
   * @default false
   */
  noUploadImg: {
    type: Boolean
  },
  /**
   * 文本区域自动获得焦点
   *
   * @default false
   */
  autoFocus: {
    type: Boolean
  },
  /**
   * 禁用文本区域
   *
   * @default false
   */
  disabled: {
    type: Boolean
  },
  /**
   * 文本区域为只读
   *
   * @default false
   */
  readOnly: {
    type: Boolean
  },
  /**
   * 文本区域允许的最大字符数
   */
  maxLength: {
    type: Number
  },
  /**
   * 是否启用自动识别粘贴代码类别
   * 目前支持 vscode 复制的代码识别
   *
   * @default false
   */
  autoDetectCode: {
    type: Boolean
  },
  /**
   * 输入框失去焦点时触发事件
   */
  onBlur: {
    type: Function
  },
  /**
   * 输入框获得焦点时触发事件
   */
  onFocus: {
    type: Function
  },
  /**
   * @codemirror/autocomplete匹配关键词的方法列表
   *
   * 它会被像下面这样嵌入编辑器
   *
   * import { autocompletion } from '@codemirror/autocomplete';
   * autocompletion({
   *   override: [...completions]
   * })
   */
  completions: {
    type: Array
  },
  /**
   * 是否在工具栏下面显示对应的文字名称
   *
   * @default false
   */
  showToolbarName: {
    type: Boolean,
    default: !1
  },
  /**
   * 字符输入事件
   */
  onInput: {
    type: Function
  }
}, ae = ["onHtmlChanged", "onGetCatalog"], wt = [
  ...ae,
  "onChange",
  "onSave",
  "onUploadImg",
  "onError",
  "update:modelValue",
  "onBlur",
  "onFocus",
  "onInput"
], Ye = /* @__PURE__ */ J({
  name: "MdPreview",
  props: re,
  emits: ae,
  setup(e, a) {
    const {
      editorId: l,
      noKatex: t,
      noMermaid: n,
      noHighlight: r
    } = e;
    return ne(e), oe(e), ce(() => {
      g.clear(l);
    }), () => H("div", {
      id: l,
      class: [h, e.class, e.theme === "dark" && `${h}-dark`, `${h}-previewOnly`],
      style: e.style
    }, [H(Xe, {
      modelValue: e.modelValue,
      onHtmlChanged: (i) => {
        e.onHtmlChanged ? e.onHtmlChanged(i) : a.emit("onHtmlChanged", i);
      },
      onGetCatalog: (i) => {
        e.onGetCatalog ? e.onGetCatalog(i) : a.emit("onGetCatalog", i);
      },
      mdHeadingId: e.mdHeadingId,
      noMermaid: n,
      sanitize: e.sanitize,
      noKatex: t,
      formatCopiedText: e.formatCopiedText,
      noHighlight: r,
      noImgZoomIn: e.noImgZoomIn,
      previewOnly: !0
    }, null)]);
  }
}), R = Ye;
R.install = (e) => (e.component(R.name, R), e);
export {
  Ie as C,
  R as M,
  Y as O,
  Se as T,
  ee as a,
  Xe as b,
  dt as c,
  wt as d,
  bt as e,
  ft as f,
  pt as g,
  ht as h,
  gt as i,
  vt as j,
  yt as k,
  mt as u
};
